---
title: "Plasmid Network Analysis"
author: "Max Cummins"
date: "2024-05-03"
output: html_document
---

```{r setup, include=FALSE}
#Read in our packages
library(tidyverse)
library(vroom)
library(ggplot2)
library(xml2)
library(ggrepel)
library(caret)
library(here)
library(igraph)

knitr::opts_chunk$set(echo = TRUE, root.dir = here::here())

#Define our not in function
`%nin%` <- Negate(`%in%`)
```

# MobSuite Processing

Below we will explore our plasmid data as generated by MobSuite.

## Read in our tree and genomic and metadata dataset

```{r Merge_analysis_outputs, include=FALSE}

genometa <- vroom("delims/genometa_n5631.txt", show_col_types = FALSE)

#Read in our tree
tree <- ggtree::read.tree("analysis/pad.nwk")

#Remove quotes from sample names
tree$tip.label <- gsub("'", "", tree$tip.label)

#Create a list of samples in our cohort
sample_names <- tree$tip.label %>% as.data.frame() %>% rename('name' = '.')

```

## Read in our plasmid cluster data

```{r}
if(!exists("plas_clusters")){
        if(file.exists("delims/plasmid_cluster_data.txt")){
           plas_clusters <- vroom("delims/plasmid_cluster_data.txt")     
        }else{source("scripts/plasmid_data_aggregate.R")}
}
```



## MobTyper Results
```{r}
#Read in our mobtyper data
mobtyper_df <- vroom("analysis/mobtyper_results_concatenated.txt", show_col_types = FALSE)

#Filter in our mobtyper data and rename the sample ID column to name
mobtyper_df <- mobtyper_df %>% rename('name' = sample_id) %>% mutate(name = str_replace(name, ":.*", "")) %>% filter(name %in% tree$tip.label)
```


## Entropy attempt
```{r}
detected_plas_clusters <- plas_clusters %>% filter(primary_cluster_id %in% unique(mobtyper_df$primary_cluster_id))


ST_entropy<- genometa %>%
        group_by(ST_new) %>%
        add_count(name = 'Count_of_ST') %>%
        ungroup() %>%
        select(name, ST_new, Count_of_ST) %>%
        inner_join(mobtyper_df, by = "name") %>%
        group_by(ST_new, primary_cluster_id) %>%
        summarise(counts = n(), Count_of_ST) %>%
        unique() %>%
        group_by(ST_new, primary_cluster_id) %>% 
        mutate(ST_cluster_prob = counts/Count_of_ST) %>%
        mutate(log_ST_cluster_prob = log(ST_cluster_prob)) %>%
        mutate(prod_ST_cluster_prob_and_log = ST_cluster_prob*log_ST_cluster_prob) %>%
        group_by(primary_cluster_id) %>%
        mutate(entropy = abs(sum(prod_ST_cluster_prob_and_log))) %>%
        mutate('ST_entropy' = entropy) %>% select(primary_cluster_id, ST_entropy) %>% unique()

Phylogroup_entropy <- genometa %>%
        group_by(Consensus_phylogroup) %>%
        add_count(name = 'Count_of_Phylogroup') %>%
        ungroup() %>%
        select(name, Consensus_phylogroup, Count_of_Phylogroup) %>%
        inner_join(mobtyper_df, by = "name") %>%
        group_by(Consensus_phylogroup, primary_cluster_id) %>%
        summarise(counts = n(), Count_of_Phylogroup) %>%
        unique() %>%
        group_by(Consensus_phylogroup, primary_cluster_id) %>% 
        mutate(Phylogroup_cluster_prob = counts/Count_of_Phylogroup) %>%
        mutate(log_Phylogroup_cluster_prob = log(Phylogroup_cluster_prob)) %>%
        mutate(prod_Phylogroup_cluster_prob_and_log = Phylogroup_cluster_prob*log_Phylogroup_cluster_prob) %>%
        group_by(primary_cluster_id) %>%
        mutate(entropy = abs(sum(prod_Phylogroup_cluster_prob_and_log))) %>%
        mutate('Phylogroup_entropy' = entropy) %>%
        select(primary_cluster_id, Phylogroup_entropy) %>% unique()

Source_entropy<- genometa %>%
        group_by(Revised_Source_Niche) %>%
        add_count(name = 'Count_of_Revised_Source_Niche') %>%
        ungroup() %>%
        select(name, Revised_Source_Niche, Count_of_Revised_Source_Niche) %>%
        inner_join(mobtyper_df, by = "name") %>%
        group_by(Revised_Source_Niche, primary_cluster_id) %>%
        summarise(counts = n(), Count_of_Revised_Source_Niche) %>%
        unique() %>%
        group_by(Revised_Source_Niche, primary_cluster_id) %>% 
        mutate(Revised_Source_Niche_cluster_prob = counts/Count_of_Revised_Source_Niche) %>%
        mutate(log_Revised_Source_Niche_cluster_prob = log(Revised_Source_Niche_cluster_prob)) %>%
        mutate(prod_Revised_Source_Niche_cluster_prob_and_log = Revised_Source_Niche_cluster_prob*log_Revised_Source_Niche_cluster_prob) %>%
        group_by(primary_cluster_id) %>%
        mutate(entropy = abs(sum(prod_Revised_Source_Niche_cluster_prob_and_log))) %>%
        mutate('Source_entropy' = entropy) %>% select(primary_cluster_id, Source_entropy) %>% unique()

detected_plas_clusters_w_entropy <- left_join(detected_plas_clusters, Phylogroup_entropy, by = "primary_cluster_id", relationship = "many-to-many") %>% left_join(., ST_entropy, by = "primary_cluster_id", relationship = "many-to-many") %>% left_join(., Source_entropy, by = "primary_cluster_id", relationship = "many-to-many") %>% select(sample_id, matches("entropy"), Plasmid_CIA_status, Plasmid_count_of_AMR_genes, Plasmid_VF_count, matches("^inc"), `IncF RST`, everything()) %>%
        mutate(Plasmid_CIA_status = replace_na(Plasmid_CIA_status, 0)) %>%
        mutate(Plasmid_count_of_AMR_genes = replace_na(Plasmid_count_of_AMR_genes, 0)) %>%
        mutate(Plasmid_VF_count = replace_na(Plasmid_VF_count, 0))

plot_df <- detected_plas_clusters_w_entropy %>%
        group_by(primary_cluster_id) %>%
        add_count(`rep_type(s)`) %>%
        #filter(primary_cluster_id %in% mobtyper_results$primary_cluster_id) %>%
        #Identify the most common replicon types within the remaning plasmids
        mutate(Major_reps = `rep_type(s)`[n == max(n)][1]) %>%
        select(-n) %>%
        add_count(`IncF RST`) %>%
        #filter(primary_cluster_id %in% mobtyper_results$primary_cluster_id) %>%
        #Identify the most common replicon types within the remaning plasmids
        mutate(Major_IncF_RST = `IncF RST`[n == max(n)][1]) %>%
        mutate(median_cluster_AMR_gene_count = median(Plasmid_count_of_AMR_genes)) %>%
        as.data.frame() %>% 
        group_by(primary_cluster_id) %>% 
        summarise(primary_cluster_id,
                  median_size = median(size),
                  Phylogroup_entropy,
                  ST_entropy,
                  Source_entropy,
                  Major_reps,
                  Major_IncF_RST,
                  median_VF_count = median(Plasmid_VF_count),
                  median_ARG_count = median(Plasmid_count_of_AMR_genes),
        ) %>%
        unique() 

plot_df <- mobtyper_df %>% group_by(primary_cluster_id) %>% add_count(name = "Total cluster members") %>% select(primary_cluster_id, `Total cluster members`) %>% inner_join(plot_df)


#plot_df %>% arrange(desc(`Total cluster members`)) %>% group_by(primary_cluster_id) %>% slice_head(n = 1) %>% ggplot() +
#        geom_point(aes(x = Phylogroup_entropy, y = median_VF_count, color = predicted_mobility)) +
#        scale_fill_manual(values = c('conjugative' = "red", 'non-conjugative' = 'blue', 'non-mobalizable' = 'green')) +
#        geom_text_repel(aes(x = Phylogroup_entropy, y = median_VF_count, label = primary_cluster_id))


```

```{r}
novel_and_plasmid_counts <- genometa %>%
        #Select columns of interest
        select(name, Revised_Source_Niche, Consensus_phylogroup) %>%
        #group by source to enable count
        group_by(Revised_Source_Niche) %>%
        #add source count
        add_count(name = "Source_Count") %>%
        #remove grouping
        ungroup() %>%
        #Bind to our mobtyper data
        full_join(mobtyper_df, by = "name") %>%
        #Group by sample
        group_by(name) %>%
        #Count the number of plasmids in a sample
        mutate(count_plasmids = n()) %>%
        #Correct where a count of zero is misidentified as a count of 1 for samples with no mobtyper data
        mutate(count_plasmids = if_else(is.na(md5) & count_plasmids == 1, 0, count_plasmids)) %>%
        #Add a column to indicate whether a particular plasmid is novel or not
        mutate(is_novel =  if_else(grepl("novel", primary_cluster_id), 1, 0)) %>%
        #Correct where a gene is indicated as having a non novel plasmid but actually has no plasmids
        mutate(is_novel = if_else(is.na(md5) & is_novel == 0, NA, is_novel)) %>%
        #Remove grouping
        ungroup()


source_wise_novel_total_plasmid_counts <- novel_and_plasmid_counts %>%
        #Group by Source and is novel to enable counts
        group_by(Revised_Source_Niche, is_novel) %>%
        #Add a count of novel plasmids within a given source
        add_count() %>%
        #Remove grouping
        ungroup() %>%
        #Isolate columns of interest
        select(Revised_Source_Niche, Source_Count, is_novel, n) %>%
        #Remove unnecessary rows
        unique()


source_wise_novel_total_plasmid_counts %>%
        #Make into wide format to produce a table showing the
        #count of novel and known primary_cluster_ids within a given source
        pivot_wider(names_from = "is_novel", values_from = n) %>%
        #Make a column showing what percentage of the plasmids in a source are novel
        mutate(source_wise_perc = scales::percent(`1`/`0`)) %>%
        #Make a column showing what percentage of the novel plasmids are from a given source
        mutate(novel_total_perc = scales::percent(`1`/732)) %>%
        #Make a column showing what percentage of the full collection is represented by a given source
        mutate(collection_total_perc = scales::percent(Source_Count/5631))  %>%
        #Make a column showing the average number of novel plasmids carried by strains within a given source
        mutate(adjusted_mean_carriage = `1`/Source_Count) %>%
        #Adjust this based on the mean number of novel plasmids observed within a given source
        mutate(num_strains_per_novel = 1/adjusted_mean_carriage) %>%
        #Make a column showing the mean number of all plasmids within a given source
        mutate(mean_total_plas = (`0`+`1`)/Source_Count) %>%
        #Make a column that penalises sources based on their mean count of total plasmids to proportionally reduce
        #their prevalence of carriage of novel plasmids
        mutate(adj_num_strains_per_novel = num_strains_per_novel/mean_total_plas) %>%
        #Select columns of interest
        select(Revised_Source_Niche,
               Source_Count,
               `0`,
               `1`,
               `NA`,
               adjusted_mean_carriage,
               mean_total_plas,
               num_strains_per_novel,
               adj_num_strains_per_novel,
               source_wise_perc,
               novel_total_perc,
               collection_total_perc) %>%
        unique()

anova_df_source_novel <- novel_and_plasmid_counts %>%
        select('name', 'Revised_Source_Niche','count_plasmids') %>%
        mutate(Revised_Source_Niche = factor(Revised_Source_Niche,
                                             levels = c("Food",
                                                        "Environmental",
                                                        "Livestock",
                                                        "Wild Animal",
                                                        "Human",
                                                        "Companion Animal"))) %>%
        unique()

set.seed(1)
subsampled_anova_df_source_novel <- anova_df_source_novel %>% group_by(Revised_Source_Niche) %>% sample_n(302)

subsampled_anova_df_source_novel %>% select(name) %>% as.data.frame() %>% write_delim("delims/302_subset_names.txt")

subsampled_anova_Source_novel <- aov(count_plasmids ~ Revised_Source_Niche, data = subsampled_anova_df_source_novel)

summary(subsampled_anova_Source_novel)

tukey_result <- TukeyHSD(subsampled_anova_Source_novel)



#model <- lm(n_novel ~ Revised_Source_Niche + count_plasmids, data = subsampled_anova_df_source_novel)

#summary(model)

#test_set <- anova_df_source_novel %>% filter(name %nin% subsampled_anova_df_source_novel$name)

#test_set$pred <- predict(object = model, test_set, type = "response")


#Source Colours
Source_cols <- c("Companion Animal" = "#59398d",
         "Environmental" = "#709b46",
         "Food" = "#c09f3d",
         "Human" = "#48c595",
         "Livestock" =  "#c26bbc",
         "Wild Animal" = "#b9553d")

#source_Novel_count_plot <- subsampled_anova_df_source_novel %>%
#        ggplot() +
#        geom_boxplot(aes(x = Revised_Source_Niche, y = n_novel, fill = Revised_Source_Niche)) +
#        scale_fill_manual(values = Source_cols)



```



```{r}

mobtyper_results <- read_delim("analysis/mobtyper_results.txt", 
    delim = "\t", escape_double = FALSE, 
    trim_ws = TRUE)

metadata <- genometa %>% select(1:20)

# Create an adjacency matrix:
chord_data_clusters <- mobtyper_results %>% filter(primary_cluster_id != "-") %>% rename('name' = sample_id) %>% inner_join(metadata, by = 'name') %>% select(name, primary_cluster_id, Revised_Source_Niche) %>%
        unique() %>%
        mutate(primary_cluster_id_other = fct_lump_n(primary_cluster_id, 50)) %>%
        mutate(primary_cluster_id = primary_cluster_id_other) %>%
        select(-primary_cluster_id_other) %>% 
        filter(Revised_Source_Niche != "Unknown") %>%
        group_by(primary_cluster_id, Revised_Source_Niche) %>%
        summarise(counts = n()) %>%
        mutate(primary_cluster_id = as.factor(primary_cluster_id)) %>% 
        pivot_wider(id_cols = primary_cluster_id, names_from = Revised_Source_Niche, values_from = counts, values_fill = 0) %>%
        column_to_rownames("primary_cluster_id") %>% as.matrix()
```


```{r}

STs <- genometa %>% select(name, ST_new)

plas_clusters_w_ST <- mobtyper_df %>% select(name, primary_cluster_id) %>% inner_join(STs, by = 'name') %>% filter(primary_cluster_id %in% rownames(chord_data_clusters))

plas_cluster_pairs <- plas_clusters_w_ST %>% rename('name2' = name, 'ST_new2' = ST_new) %>% inner_join(plas_clusters_w_ST, by = 'primary_cluster_id')

plas_cluster_pairs <- plas_cluster_pairs %>% filter(name != name2)

```

```{r}

#igraph_input <- igraph_input %>% filter(primary_cluster_id == "AA161")

for(primary_clust in unique(sort(plas_cluster_pairs$primary_cluster_id))){
  
  set.seed(1)
  
  #Extract sources from genotypic/metadata table
  sources <- genometa %>% select(name, Revised_Source_Niche, ST_new, Consensus_phylogroup)
  
  #Set a colour for each source
  sources <- sources %>%
    mutate(color = case_when(
      #Revised_Source_Niche == "Captive Animal" ~ "#b5467395",
      Revised_Source_Niche == "Companion Animal" ~ "#59398d95",
      Revised_Source_Niche == "Environmental" ~ "#709b4695",
      Revised_Source_Niche == "Food" ~ "#c09f3d95",
      Revised_Source_Niche == "Human" ~ "#48c59595",
      Revised_Source_Niche == "Livestock" ~ "#c26bbc95",
      #Revised_Source_Niche == "Waste" ~ "#6d83da95",
      Revised_Source_Niche == "Wild Animal" ~ "#b9553d95"
    ))
  
  g_df <- plas_cluster_pairs %>% filter(primary_cluster_id == primary_clust) %>% filter(ST_new2 == ST_new) %>% select(-primary_cluster_id, -ST_new, -ST_new2)
  
  # Convert data frame to edge list for igraph
  g <- graph_from_data_frame(d = g_df, directed = FALSE)
  
  g <- igraph::simplify(g)
  
  E(g)$weight <- as_edgelist(g, names = T) %>%
    as.data.frame() %>%
    rename("name" = V1) %>%
    left_join(sources, by = "name") %>%
    rename("V1" = name) %>%
    rename("name" = V2) %>%
    left_join(sources, by = "name") %>%
    select(name, V1, Revised_Source_Niche.x, Revised_Source_Niche.y) %>%
    mutate(weight = if_else(Revised_Source_Niche.x == Revised_Source_Niche.y, 10, 1)) %>%
    pull(weight)
  
  #Get a list of vertex names
  g_vertex_names <- V(g)$name 
  
  # Match colors to vertices based on names
  vertex_color_vector <- sources$color[match(g_vertex_names, sources$name)]
  
  set.seed(1)
  
  #Plot the graph
  pdf(file = paste0("figures/Network_graphs/Network_by_source_",
                    primary_clust,
                    ".pdf"),
      width = 8,
      height = 8)
  
  lo <- layout_with_fr(g) # create a layout
  
  max_coords <- max(abs(lo))
  layout <- lo / max_coords * 0.5  # Adjust scale factor to spread out the layout more
  
  
  #lo <- norm_coords(lo, ymin=-1, ymax=1, xmin=-1, xmax=1)
  
  # Match STs to vertices based on names
  V(g)$ST <- paste0("ST",sources$ST_new[match(g_vertex_names, sources$name)], ":", sources$Consensus_phylogroup[match(g_vertex_names, sources$name)])
  
  #Identify duplicate ST labels
  
  duplicates <- duplicated(V(g)$ST)
  
  V(g)$ST[duplicates] <- NA
  
  layout <- layout_with_fr(g, weights = E(g)$weight)
  
  mode <- function(x) { names(which.max(table(x))) } 
  
  plasmid_clusters_reduced <- detected_plas_clusters_w_entropy %>%
    mutate(ColV = if_else(is.na(ColV), 0, ColV)) %>% 
    mutate(ABRICATE..vfdb_senB = if_else(is.na(ABRICATE..vfdb_senB), 0, ABRICATE..vfdb_senB)) %>% 
    group_by(primary_cluster_id) %>%
    summarise(median_ARG = median(Plasmid_count_of_AMR_genes, drop.NA = TRUE), median_VF = median(Plasmid_VF_count, drop.NA = TRUE), common_RST = mode(`IncF RST`), ColV = mode(ColV), senB = mode(ABRICATE..vfdb_senB))
  
  plasmid_clusters_reduced <- plasmid_clusters_reduced %>% mutate(fig_main = paste(primary_cluster_id, common_RST , "ARG:", median_ARG, "VF:", median_VF, "ColV:", ColV, "SenB:", senB)) %>% filter(primary_cluster_id == primary_clust)
  
  set.seed(1)
  
  plot(g,
       edge.width = E(g)$thickness,
       rescale = TRUE,
       main = plasmid_clusters_reduced$fig_main[1], # mark.groups = paste0("ST", 1:52),
       layout = layout,
       edge.color = "grey80",
       vertex.size = 2.5,
       vertex.label = V(g)$ST,
       vertex.label.cex = 0.5,
       vertex.label.dist = 1,
       vertex.color = vertex_color_vector,
       asp = 0)
  
  # legend("bottomleft",
  #        legend = c(#"Captive Animal",
  #            "Companion Animal", 
  #            "Environmental", 
  #            "Food", 
  #            "Human", 
  #            "Livestock", 
  #            #"Waste", 
  #            "Wild Animal"),
  #        fill = c(
  #          #"#b5467395",
  #          "#59398d95",
  #          "#709b4695",
  #          "#c09f3d95",
  #          "#48c59595",
  #          "#c26bbc95",
  #          #"#6d83da95",
  #          "#b9553d95"),
  #        pch=19,
  #        bty = "n",
  #        pt.cex = 1,
  #        cex = 1,
  #        text.col="black" ,
  #        horiz = FALSE)
  
  dev.off()
}

```



```{r}

primary_clust <- "AA176"

#igraph_input <- igraph_input %>% filter(primary_cluster_id == "AA161")
set.seed(1)

#Extract sources from genotypic/metadata table
sources <- genometa %>% select(name, Revised_Source_Niche, ST_new)

#Set a colour for each source
sources <- sources %>%
  mutate(color = case_when(
    #Revised_Source_Niche == "Captive Animal" ~ "#b5467395",
    Revised_Source_Niche == "Companion Animal" ~ "#59398d95",
    Revised_Source_Niche == "Environmental" ~ "#709b4695",
    Revised_Source_Niche == "Food" ~ "#c09f3d95",
    Revised_Source_Niche == "Human" ~ "#48c59595",
    Revised_Source_Niche == "Livestock" ~ "#c26bbc95",
    #Revised_Source_Niche == "Waste" ~ "#6d83da95",
    Revised_Source_Niche == "Wild Animal" ~ "#b9553d95"
  ))


g_df <- plas_cluster_pairs %>% filter(primary_cluster_id == primary_clust) %>% filter(ST_new2 == ST_new) %>% select(-primary_cluster_id, -ST_new, -ST_new2)

# Convert data frame to edge list for igraph
g <- graph_from_data_frame(d = g_df, directed = FALSE)

g <- igraph::simplify(g)

#Get a list of vertex names
g_vertex_names <- V(g)$name 

# Match colors to vertices based on names
vertex_color_vector <- sources$color[match(g_vertex_names, sources$name)]

set.seed(1)

# Plot the graph
#pdf(file = paste0("figures/Network_graphs/Network_by_source_",
#                  primary_clust,
#                  ".pdf"),
#    width = 11.69,
#    height = 8.27)

lo <- layout_with_fr(g) # create a layout

max_coords <- max(abs(lo))
layout <- lo / max_coords * 0.5  # Adjust scale factor to spread out the layout more


#lo <- norm_coords(lo, ymin=-1, ymax=1, xmin=-1, xmax=1)

# Match STs to vertices based on names
V(g)$ST <- paste0("ST",sources$ST_new[match(g_vertex_names, sources$name)])

#Identify duplicate ST labels

duplicates <- duplicated(V(g)$ST)

V(g)$ST[duplicates] <- NA


set.seed(1)

plot(g,
     #edge.width = E(g)$thickness,
     rescale = TRUE,
     main = primary_clust,# mark.groups = paste0("ST", 1:52),
     layout = layout_with_fr,
     edge.color = "grey80",
     vertex.size = 1,
     vertex.label = V(g)$ST,
     vertex.label.cex = 0.5,
     vertex.label.dist = 1,
     vertex.color = vertex_color_vector,
     asp = 0)

legend("bottomleft",
       legend = c(#"Captive Animal",
           "Companion Animal", 
           "Environmental", 
           "Food", 
           "Human", 
           "Livestock", 
           #"Waste", 
           "Wild Animal"),
       fill = c(
         #"#b5467395",
         "#59398d95",
         "#709b4695",
         "#c09f3d95",
         "#48c59595",
         "#c26bbc95",
         #"#6d83da95",
         "#b9553d95"),
       pch=19,
       bty = "n",
       pt.cex = 1,
       cex = 1,
       text.col="black" ,
       horiz = FALSE)

dev.off()

```





