---
title: "Figure_5"
author: "Max Cummins"
date: "`r Sys.Date()`"
output:
    html_document:
        theme: spacelab
        toc: true
        toc_float: true
---

#

```{r setup, include=FALSE}
#Set wd to directory containing current markdown script
knitr::opts_chunk$set(dirname(rstudioapi::getSourceEditorContext()$path))


#Read in our packages
library(tidyverse)
library(ggplot2)
library(xml2)
library(ggtree)
library(paletteer)
library(vroom)
library(ggpubr)


#Define our not in function
`%nin%` <- Negate(`%in%`)
```

## Read in our necessary files

```{r}

#Read in our genometa sheet generated by AusTrakka_TEA_data_generation.Rmd
genometa <- vroom("../delims/genometa_n5631.txt", show_col_types = FALSE)

#Read in our tree
tree <- ggtree::read.tree("../analysis/pad.nwk")

#Remove quotes from sample names
tree$tip.label <- gsub("'", "", tree$tip.label)

#Create a list of samples in our cohort
sample_names <- tree$tip.label %>% as.data.frame() %>% rename('name' = '.')
```

## Define any variables

```{r}
#Overwrite with different colours
source_cols <- c("Companion Animal" = "#59398d",
         "Environmental" = "#709b46",
         "Food" = "#c09f3d",
         "Human" = "#48c595",
         "Livestock" =  "#c26bbc",
         "Wild Animal" = "#b9553d")
```

## Generate figure
```{r}

df_all_ST = data.frame()

for(i in list.dirs("../analysis/SKA")){
        if(str_detect(string = i, pattern = "_dists")){
                #ST extraction
                ST <- gsub("_dists", "", basename(i))
                #Read in SNP distances
                dists <- read_delim(paste0(i,"/",ST,"_dist_100_SNP.distances.tsv"), 
                    delim = "\t", escape_double = FALSE, 
                    trim_ws = TRUE, show_col_types = FALSE)
                
                #Read in SNP Clusters
                clusts <- read_delim(paste0(i,"/",ST,"_dist_100_SNP.clusters.tsv"), 
                    delim = "\t", escape_double = FALSE, 
                    trim_ws = TRUE, show_col_types = FALSE)
                
                
                #Combine our metadata and our clusters for use in microreact
                data_microreact <- left_join(clusts, genometa, by = c("ID" ="name")) %>% select(ID, starts_with("Revised_Source"), Collection_Year, Collection, fimtyper..Fimtype, ColV, Cluster__autocolour)
                
                data_microreact1 <- data_microreact %>% rename("Sample 1" = "ID") %>% rev()
                
                data_microreact2 <- data_microreact %>% rename("Sample 2" = "ID")
                
                dists2 <- inner_join(data_microreact1, dists, by = "Sample 1")
                
                dists2 <- inner_join(dists2, data_microreact2, by = "Sample 2")
                
                dists2 <- dists2 %>% mutate(across(where(is.factor), as.character))
                
                dists_w_metadata <- dists2 %>% rowwise() %>%
                        mutate(col1 = pmin(Revised_Source_Niche.x, Revised_Source_Niche.y), col2 = pmax(Revised_Source_Niche.x, Revised_Source_Niche.y), source_combo = paste0(col1, "/",col2, collapse = "")) %>% select(-col1, -col2)
                
                dists_w_metadata <- dists_w_metadata %>% rowwise() %>%
                        mutate(col1 = pmin(Revised_Source_Type.x, Revised_Source_Type.y), col2 = pmax(Revised_Source_Type.x, Revised_Source_Type.y), source_type_combo = paste0(col1, "/",col2, collapse = "")) %>% select(-col1, -col2)
                
                dists_w_metadata$ST.x <- ST
                
                df_all_ST <- rbind(df_all_ST, dists_w_metadata)
                
               # write_delim(x = data_microreact1, file = paste0(i,"/",ST,"_metadata_microreact.tsv"), delim = "\t")
        }
        
}
```

```{r}
#library(harrietr)

df_all_ST <- df_all_ST %>% as.data.frame() %>% filter(`Sample 1` %in% tree$tip.label & `Sample 2` %in% tree$tip.label)

dists_w_metadata_x_source <- df_all_ST %>% as.data.frame() %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y)

cgMLST_clusters <- read_delim(file = "../analysis/clusters.txt")

if(!file.exists("../analysis/pad.txt")){
  unzip(exdir = "../analysis", zipfile = "../analysis/pad.txt.zip", overwrite = FALSE)
}

allelic_distances <- read_delim(file = "../analysis/pad.txt", show_col_types = FALSE)

allelic_distances <- allelic_distances %>% rename(name = '5631')

long_dists <- pivot_longer(allelic_distances, cols = 2:ncol(allelic_distances), names_repair = 'minimal')

colnames(long_dists) <- c('Sample 1', 'Sample 2', 'value')

distsnew <- inner_join(df_all_ST, long_dists)

dists_w_metadata_all_source <- distsnew %>% as.data.frame() 

df_all_ST_filtered_SNPs_all <- dists_w_metadata_all_source %>% as.data.frame() 

pal <- colorRampPalette(c("red", "yellow", "blue"))
dot_cols <- pal(41)

df_all_ST_filtered_SNPs_all %>%
        mutate(source_combo = str_replace(source_combo, 'Companion Animal', 'Comp. Animal')) %>% 
        mutate(source_combo = str_replace(source_combo, 'Environmental', 'Environ.')) %>% 
        filter(Revised_Source_Niche.x != Revised_Source_Niche.y) %>% 
        filter(value < 40)  %>% 
        mutate(value = as.factor(value)) %>%
        rename('Allelic\nDistance\n(cgMLST)' = value) %>%
        rename('Sequence type' = ST.x) %>%
        ggplot() +
        geom_boxplot(aes(x = SNPs, y = `Sequence type`)) +
        geom_jitter(aes(x = SNPs, y = `Sequence type`, color = `Allelic\nDistance\n(cgMLST)`), size = 0.25) +
        scale_color_manual(values = dot_cols, breaks = as.factor(0:40)) +
        facet_wrap(~source_combo, nrow = 7, ncol = 2, strip.position="right") +
        theme(axis.text.x = element_text(angle = 90)) +
        geom_vline(xintercept = 100, color = "red", linetype = "dashed") +
        scale_x_continuous(breaks = seq(0,1300, by = 100)) +
        guides(colour = guide_legend(override.aes = list(size=1.5), ncol = 1))
```

Supplementary Figure X
```{r}
plot1 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs < 12000) %>%
        mutate(fimtyper..Fimtype.x = if_else(is.na(fimtyper..Fimtype.x), "Missing", fimtyper..Fimtype.x)) %>%
        mutate(fimtyper..Fimtype.y = if_else(is.na(fimtyper..Fimtype.y), "Missing", fimtyper..Fimtype.y)) %>%
        mutate(same_fim = if_else(fimtyper..Fimtype.x == fimtyper..Fimtype.y, T, F)) %>% 
        ggplot(aes(x = SNPs, y = value)) + #, color=intra_inter
        geom_point(aes(color = same_fim)) +
        facet_wrap(~ST.x) +
        labs(x = "Pairwise SNP Distance", y = "Pairwise Allelic Distance") +
        geom_smooth(
                color = "red",
                fill = "#69b3a2",
                se = TRUE
        ) +
        stat_cor(method = "spearman", aes(label = paste(..rr.label.., ..p.label.., sep = "~")), color = "red", geom = "label")# +

ggsave(filename = "../Figures/SNPs_vs_cgMLST_facet_ST_w_fim.pdf", plot = plot1, width = 11.69, height = 8.27, units = "in", dpi = 300)
```

Supplementary Figure X
```{r}

STs <- df_all_ST_filtered_SNPs_all$ST.x %>% unique()

ST_cols <- paletteer_d("ggthemes::Tableau_10") %>% as.vector()

names(ST_cols) <- STs


plot2 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs < 12000) %>%
        mutate(fimtyper..Fimtype.x = if_else(is.na(fimtyper..Fimtype.x), "Missing", fimtyper..Fimtype.x)) %>%
        mutate(fimtyper..Fimtype.y = if_else(is.na(fimtyper..Fimtype.y), "Missing", fimtyper..Fimtype.y)) %>%
        mutate(same_fim = if_else(fimtyper..Fimtype.x == fimtyper..Fimtype.y, T, F)) %>% 
        ggplot(aes(x = SNPs, y = value)) + #, color=intra_inter
        geom_point(aes(color = ST.x), alpha = I(0.2)) +
        labs(x = "Pairwise SNP Distance", y = "Pairwise Allelic Distance") +
        geom_smooth(
                color = "red",
                fill = "#69b3a2",
                se = TRUE
        ) +
        stat_cor(method = "spearman", aes(label = paste(..rr.label.., ..p.label.., sep = "~")), color = "red", geom = "label") +
        scale_color_manual(values = ST_cols)

ggsave(filename = "../Figures/SNPs_vs_cgMLST_col_by_ST.pdf", plot = plot2, width = 11.69, height = 8.27, units = "in", dpi = 300)

```
```{r}
plot3 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs < 100) %>%
        mutate(fimtyper..Fimtype.x = if_else(is.na(fimtyper..Fimtype.x), "Missing", fimtyper..Fimtype.x)) %>%
        mutate(fimtyper..Fimtype.y = if_else(is.na(fimtyper..Fimtype.y), "Missing", fimtyper..Fimtype.y)) %>%
        mutate(same_fim = if_else(fimtyper..Fimtype.x == fimtyper..Fimtype.y, T, F)) %>% 
        ggplot(aes(x = SNPs, y = value)) + #, color=intra_inter
        geom_point(aes(color = ST.x), alpha = I(0.4)) +
        labs(x = "Pairwise SNP Distance", y = "Pairwise Allelic Distance") +
        geom_smooth(
                color = "red",
                fill = "#69b3a2",
                se = TRUE
        ) +
        stat_cor(method = "spearman", aes(label = paste(..rr.label.., ..p.label.., sep = "~")), color = "red", geom = "label") +
        scale_color_manual(values = ST_cols)

ggsave(filename = "../Figures/SNPs_vs_cgMLST_col_by_ST_sub_100.pdf", plot = plot3, width = 11.69, height = 8.27, units = "in", dpi = 300)

plot4 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs < 100) %>%
        mutate(fimtyper..Fimtype.x = if_else(is.na(fimtyper..Fimtype.x), "Missing", fimtyper..Fimtype.x)) %>%
        mutate(fimtyper..Fimtype.y = if_else(is.na(fimtyper..Fimtype.y), "Missing", fimtyper..Fimtype.y)) %>%
        mutate(same_fim = if_else(fimtyper..Fimtype.x == fimtyper..Fimtype.y, T, F)) %>% 
        ggplot(aes(x = SNPs, y = value)) + #, color=intra_inter
        geom_point(aes(color = same_fim), alpha = I(0.2)) +
        facet_wrap(~ST.x) +
        labs(x = "Pairwise SNP Distance", y = "Pairwise Allelic Distance") +
        geom_smooth(
                color = "red",
                fill = "#69b3a2",
                se = TRUE
        ) +
        stat_cor(method = "spearman", aes(label = paste(..rr.label.., ..p.label.., sep = "~")), color = "red", geom = "label")

ggsave(filename = "../Figures/SNPs_vs_cgMLST_facet_ST_w_fim_sub_100.pdf", plot = plot4, width = 11.69, height = 8.27, units = "in", dpi = 300)
```

```{r}
plot5 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs < 12000) %>%
        mutate(fimtyper..Fimtype.x = if_else(is.na(fimtyper..Fimtype.x), "Missing", fimtyper..Fimtype.x)) %>%
        mutate(fimtyper..Fimtype.y = if_else(is.na(fimtyper..Fimtype.y), "Missing", fimtyper..Fimtype.y)) %>%
        mutate(same_source = if_else(Revised_Source_Niche.x == Revised_Source_Niche.y, "Intersource", "Intrasource")) %>% 
        ggplot(aes(x = SNPs, y = value)) + #, color=intra_inter
        geom_point(aes(color = ST.x), alpha = I(0.4)) +
        facet_wrap(~same_source) +
        labs(x = "Pairwise SNP Distance", y = "Pairwise Allelic Distance") +
        geom_smooth(
                method = "glm",
                color = "red",
                fill = "#69b3a2",
                se = TRUE
        ) +
        stat_cor(method = "spearman") +
        scale_color_manual(values = ST_cols)

ggsave(filename = "../Figures/SNPs_vs_cgMLST_facet_same_source_col_by_ST.pdf", plot = plot5, width = 11.69, height = 8.27, units = "in", dpi = 300)
```


## Euler Diagram

```{r}
#Install eulerr if not already present
#install.packages("eulerr")

library(eulerr)

set.seed(5) 

intersource_less_20 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs <= 20) %>% filter(`Sample 1` != `Sample 2`) %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y)

intersource_less_20 <- intersource_less_20 %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)



intersource_less_50 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs <= 50) %>% filter(`Sample 1` != `Sample 2`) %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y)

intersource_less_50 <- intersource_less_50 %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)




intersource_less_75 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs <= 75) %>% filter(`Sample 1` != `Sample 2`) %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y)

intersource_less_75 <- intersource_less_75 %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)



intersource_less_100 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs <= 100) %>% filter(`Sample 1` != `Sample 2`) %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y)

intersource_less_100 <- intersource_less_100 %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)





intersource_less_200 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs <= 200) %>% filter(`Sample 1` != `Sample 2`) %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y)

intersource_less_200 <- intersource_less_200 %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)





intersource_less_500 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs <= 500) %>% filter(`Sample 1` != `Sample 2`) %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y)

intersource_less_500 <- intersource_less_500 %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)





intersource_less_1000 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs <= 1000) %>% filter(`Sample 1` != `Sample 2`) %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y)

intersource_less_1000 <- intersource_less_1000 %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)




intersource_less_2500 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs <= 2500) %>% filter(`Sample 1` != `Sample 2`) %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y)

intersource_less_2500 <- intersource_less_2000 %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)





intersource_less_5000 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs <= 5000) %>% filter(`Sample 1` != `Sample 2`) %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y)

intersource_less_5000 <- intersource_less_2000 %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)


intersource_less_7500 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs <= 7500) %>% filter(`Sample 1` != `Sample 2`) %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y)

intersource_less_7500 <- intersource_less_7500 %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)



intersource_less_10000 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs <= 10000) %>% filter(`Sample 1` != `Sample 2`) %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y)

intersource_less_10000 <- intersource_less_10000 %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)



intersource_total_11350 <- df_all_ST_filtered_SNPs_all %>% filter(`Sample 1` != `Sample 2`) %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y)

intersource_total_11350 <- intersource_total_11350 %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)



sets <- list('20' = intersource_less_20$name_combo,
             '50' = intersource_less_50$name_combo,
             '75' = intersource_less_75$name_combo,
             '100' = intersource_less_100$name_combo,
             '200' = intersource_less_200$name_combo,
             '500' = intersource_less_500$name_combo,
             '1000' = intersource_less_1000$name_combo,
             '2500' = intersource_less_2500$name_combo,
             '5000' = intersource_less_5000$name_combo,
             '7500' = intersource_less_7500$name_combo,
             '11,350' = intersource_total_11350$name_combo)

# Generate and plot the Euler diagram
fit <- euler(sets)
plot(fit,
     quantities = TRUE,
     lty = c(1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2),
     fills = rev(
       c("#FFFFF0",
         "grey24",
         "grey36",
         "grey48",
         "grey60",
         "grey72",
         "grey84",
         "#FF6A6A",
         "#FFA500",
         "#FFFF00",
         "#FFFFF0")))

```

## Euler Diagram

```{r}
#Install eulerr if not already present
#install.packages("eulerr")

set.seed(5)

intrasource_less_20 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs <= 20) %>% filter(`Sample 1` != `Sample 2`) %>% filter(Revised_Source_Niche.x == Revised_Source_Niche.y)

intrasource_less_20 <- intrasource_less_20 %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)



intrasource_less_50 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs <= 50) %>% filter(`Sample 1` != `Sample 2`) %>% filter(Revised_Source_Niche.x == Revised_Source_Niche.y)

intrasource_less_50 <- intrasource_less_50 %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)




intrasource_less_75 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs <= 75) %>% filter(`Sample 1` != `Sample 2`) %>% filter(Revised_Source_Niche.x == Revised_Source_Niche.y)

intrasource_less_75 <- intrasource_less_75 %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)


intrasource_less_100 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs <= 100) %>% filter(`Sample 1` != `Sample 2`) %>% filter(Revised_Source_Niche.x == Revised_Source_Niche.y)

intrasource_less_100 <- intrasource_less_100 %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)




intrasource_less_200 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs <= 200) %>% filter(`Sample 1` != `Sample 2`) %>% filter(Revised_Source_Niche.x == Revised_Source_Niche.y)

intrasource_less_200 <- intrasource_less_200 %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)




intrasource_less_500 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs <= 500) %>% filter(`Sample 1` != `Sample 2`) %>% filter(Revised_Source_Niche.x == Revised_Source_Niche.y)

intrasource_less_500 <- intrasource_less_500 %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)



sets_2 <- list('≤ 20' = intrasource_less_20$name_combo,
             '≤ 50' = intrasource_less_50$name_combo,
             '≤ 75' = intrasource_less_75$name_combo,
             '≤ 100' = intrasource_less_100$name_combo,
             '≤ 200' = intrasource_less_200$name_combo,
             '≤ 500' = intrasource_less_500$name_combo)

# Generate and plot the Euler diagram
fit_2 <- euler(sets_2)
plot(fit_2,
     quantities = TRUE,
     lty = c(1, 1, 1, 1, 2, 2),
     fills = rev(
       c("#A9A9A9",
         "#E8E8E8",
         "#FF6A6A",
         "#FFA500",
         "#FFFF00",
         "#FFFFF0")))
```

## Euler Diagram

```{r}
#Install eulerr if not already present
#install.packages("eulerr")

set.seed(5)

total_less_20 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs <= 20) %>% filter(`Sample 1` != `Sample 2`) 

total_less_20 <- total_less_20 %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)



total_less_50 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs <= 50) %>% filter(`Sample 1` != `Sample 2`) 

total_less_50 <- total_less_50 %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)




total_less_75 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs <= 75) %>% filter(`Sample 1` != `Sample 2`) 

total_less_75 <- total_less_75 %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)


total_less_100 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs <= 100) %>% filter(`Sample 1` != `Sample 2`) 

total_less_100 <- total_less_100 %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)




total_less_200 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs <= 200) %>% filter(`Sample 1` != `Sample 2`) 

total_less_200 <- total_less_200 %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)




total_less_500 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs <= 500) %>% filter(`Sample 1` != `Sample 2`) 

total_less_500 <- total_less_500 %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)



sets_3 <- list('≤ 20' = total_less_20$name_combo,
             '≤ 50' = total_less_50$name_combo,
             '≤ 75' = total_less_75$name_combo,
             '≤ 100' = total_less_100$name_combo,
             '≤ 200' = total_less_200$name_combo,
             '≤ 500' = total_less_500$name_combo)

# Generate and plot the Euler diagram
fit_3 <- euler(sets_3)
plot(fit_3,
     quantities = TRUE,
     lty = c(1, 1, 1, 1, 2, 2),
     fills = rev(
       c("#A9A9A9",
         "#E8E8E8",
         "#FF6A6A",
         "#FFA500",
         "#FFFF00",
         "#FFFFF0")))
```


## Euler Diagram

```{r}
#Install eulerr if not already present
#install.packages("eulerr")

set.seed(5)

total_less_100_SNPs <- total_less_11350_SNPs <- df_all_ST_filtered_SNPs_all %>% filter(SNPs <= 100)  %>% filter(Revised_Source_Niche.x == Revised_Source_Niche.y) %>% filter(`Sample 1` != `Sample 2`) 

total_less_100_SNPs <- total_less_100_SNPs %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2) %>% filter(SNPs <= 11350)  %>% filter(Revised_Source_Niche.x == Revised_Source_Niche.y) %>% filter(`Sample 1` != `Sample 2`) 



total_less_11350_SNPs <- total_less_11350_SNPs %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)


total_less_11350_SNPs <- df_all_ST_filtered_SNPs_all %>% filter(SNPs <= 11350)  %>% filter(Revised_Source_Niche.x == Revised_Source_Niche.y) %>% filter(`Sample 1` != `Sample 2`) 

total_less_11350_SNPs <- total_less_11350_SNPs %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)



total_less_40_alleles <- df_all_ST_filtered_SNPs_all %>% filter(value <= 40)  %>% filter(Revised_Source_Niche.x == Revised_Source_Niche.y) %>% filter(`Sample 1` != `Sample 2`) 

total_less_40_alleles <- total_less_40_alleles %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)



sets_4 <- list('≤ 100 SNPs' = total_less_40_alleles_500_SNPs$name_combo,
               '≤ 40 Alleles' = total_less_40_alleles$name_combo
               #'≤ 13500 SNPs' = total_less_11350_SNPs$name_combo
               )

# Generate and plot the Euler diagram
fit_4 <- euler(sets_4)
plot(fit_4,
     quantities = TRUE,
     #lty = c(1, 1, 1, 1, 2, 2),
     fills = rev(
       c("#A9A9A9",
#         "#E8E8E8",
#         "#FF6A6A",
         "#FFA500",
#         "#FFFF00",
         "#FFFFF0")))
```

```{r}
#Install eulerr if not already present
#install.packages("eulerr")

library(eulerr)

set.seed(5) 

intersource_less_5_cgMLST <- df_all_ST_filtered_SNPs_all %>% filter(value <= 5) %>% filter(`Sample 1` != `Sample 2`) %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y)

intersource_less_5_cgMLST <- intersource_less_5_cgMLST %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)


intersource_less_10_cgMLST <- df_all_ST_filtered_SNPs_all %>% filter(value <= 10) %>% filter(`Sample 1` != `Sample 2`) %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y)

intersource_less_10_cgMLST <- intersource_less_10_cgMLST %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)



intersource_less_20_cgMLST <- df_all_ST_filtered_SNPs_all %>% filter(value <= 20) %>% filter(`Sample 1` != `Sample 2`) %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y)

intersource_less_20_cgMLST <- intersource_less_20_cgMLST %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)



intersource_less_40_cgMLST <- df_all_ST_filtered_SNPs_all %>% filter(value <= 40) %>% filter(`Sample 1` != `Sample 2`) %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y)

intersource_less_40_cgMLST <- intersource_less_40_cgMLST %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)


intersource_less_60_cgMLST <- df_all_ST_filtered_SNPs_all %>% filter(value <= 60) %>% filter(`Sample 1` != `Sample 2`) %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y)

intersource_less_60_cgMLST <- intersource_less_60_cgMLST %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)



intersource_less_100_cgMLST <- df_all_ST_filtered_SNPs_all %>% filter(value <= 100) %>% filter(`Sample 1` != `Sample 2`) %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y)

intersource_less_100_cgMLST <- intersource_less_100_cgMLST %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)


intersource_less_250_cgMLST <- df_all_ST_filtered_SNPs_all %>% filter(value <= 250) %>% filter(`Sample 1` != `Sample 2`) %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y)

intersource_less_250_cgMLST <- intersource_less_250_cgMLST %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)




intersource_less_500_cgMLST <- df_all_ST_filtered_SNPs_all %>% filter(value <= 500) %>% filter(`Sample 1` != `Sample 2`) %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y)

intersource_less_500_cgMLST <- intersource_less_500_cgMLST %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)


intersource_less_1005_cgMLST <- df_all_ST_filtered_SNPs_all %>% filter(value <= 1005) %>% filter(`Sample 1` != `Sample 2`) %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y)

intersource_less_1005_cgMLST <- intersource_less_1005_cgMLST %>% mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>% ungroup() %>% mutate(name_combo = paste0(col1, "/",col2)) %>% select(-col1, -col2)



sets_allelic <- list('5' = intersource_less_5_cgMLST$name_combo,
             '20' = intersource_less_20_cgMLST$name_combo,
             '40' = intersource_less_40_cgMLST$name_combo,
             '100' = intersource_less_60_cgMLST$name_combo,
             '250' = intersource_less_250_cgMLST$name_combo,
             '500' = intersource_less_500_cgMLST$name_combo,
             '1005' = intersource_less_1005_cgMLST$name_combo)

# Generate and plot the Euler diagram
fit_allelic <- euler(sets_allelic)
plot(fit_allelic,
     quantities = TRUE,
     lty = c(1, 1, 1, 1, 2, 2),
     fills = rev(
       c('grey30',
         "#A9A9A9",
         "#E8E8E8",
         "#FF6A6A",
         "#FFA500",
         "#FFFF00",
         "#FFFFF0")))

```



Cluster analysis

```{r}
df_all_ST_filtered_SNPs_all %>% filter(`Sample 1` != `Sample 2`) %>% group_by(Revised_Source_Niche.x == Revised_Source_Niche.y) %>% tally()

```
### Total clusters

```{r}
pre_clust_df <- df_all_ST_filtered_SNPs_all %>%
  mutate(col1 = pmin(`Sample 1`, `Sample 2`), col2 = pmax(`Sample 1`, `Sample 2`)) %>%
  mutate(name_combo = paste0(col1, "/",col2)) %>%
  select(-col1, -col2)

pre_clust_df <- pre_clust_df %>%
  mutate("tmp" = `Sample 1`, "tmp2" = `Sample 2`) %>%
  select(-`Sample 1`, -`Sample 2`) %>%
  rename(`Sample 2` = tmp, `Sample 1` = tmp2) %>%
  bind_rows(pre_clust_df)

average <- function(...) {
  # Capture all arguments as a vector
  args <- c(...)
  
  # Calculate the sum of all arguments
  sum_args <- sum(args)
  
  # Count the number of arguments
  num_args <- length(args)
  
  # Calculate the average
  average <- sum_args / num_args
  
  # Return the average
  return(average)
}


df_clust <- pre_clust_df %>%
  filter(SNPs <= 100) %>%
  #filter(Revised_Source_Niche.x == Revised_Source_Niche.y) %>%
  filter(`Sample 1` != `Sample 2`) %>% select(`Sample 1`, `Sample 2`, SNPs) %>%
  pivot_wider(names_from = `Sample 2`, values_from = SNPs, values_fill = 0, values_fn = average) %>% column_to_rownames("Sample 1") %>% select(all_of(rownames(.)))

clust_input <- dist(df_clust)


hc <- hclust(clust_input)

cutoff_100 <- cutree(hc, h = 10) %>% as.data.frame() %>% rownames_to_column('name')

cutoff_100 <- genometa %>% select(name, ST_new, Revised_Source_Niche) %>% inner_join(cutoff_100, by = 'name') %>% rename("Cluster" = ".")

Counts_per_clust_per_ST <- cutoff_100 %>% group_by(ST_new, Cluster) %>% summarise(Counts = n())

Counts_per_source_per_clust <- cutoff_100 %>%
  group_by(Revised_Source_Niche, Cluster) %>%
  summarise(Counts = n()) %>%
  pivot_wider(names_from = Revised_Source_Niche, values_from = Counts, values_fill = 0) %>% 
  mutate(Multi_sector = `Companion Animal` + `Environmental` + `Food` + `Human` + `Livestock` + `Wild Animal`)

Table_base <- Counts_per_clust_per_ST %>% full_join(Counts_per_source_per_clust, by = "Cluster") %>% select(ST_new, Cluster, Counts, Multi_sector)

Table_base %>% filter(Counts > 1)  %>% group_by(ST_new) %>% tally(sort = T)

```



```{r}

#Read in our genometa sheet generated by AusTrakka_TEA_data_generation.Rmd
genometa <- vroom("../delims/genometa_n5631.txt", show_col_types = FALSE)


clusts <- read_delim(file = "../analysis/SKA/ST95_dists/ST95_dist_100_SNP.clusters.tsv", delim = "\t")

clusts <- clusts %>% mutate(ST = "ST95")

clusts2 <- clusts

clusts <- read_delim(file = "../analysis/SKA/ST131_dists/ST131_dist_100_SNP.clusters.tsv", delim = "\t")

clusts <- clusts %>% mutate(ST = "ST131")

clusts2 <- bind_rows(clusts2, clusts)

clusts <- read_delim(file = "../analysis/SKA/ST117_dists/ST117_dist_100_SNP.clusters.tsv", delim = "\t")

clusts <- clusts %>% mutate(ST = "ST117")

clusts2 <- bind_rows(clusts2, clusts)
clusts <- read_delim(file = "../analysis/SKA/ST963_dists/ST963_dist_100_SNP.clusters.tsv", delim = "\t")

clusts <- clusts %>% mutate(ST = "ST963")

clusts2 <- bind_rows(clusts2, clusts)
clusts <- read_delim(file = "../analysis/SKA/ST69_dists/ST69_dist_100_SNP.clusters.tsv", delim = "\t")

clusts <- clusts %>% mutate(ST = "ST69")

clusts2 <- bind_rows(clusts2, clusts)
clusts <- read_delim(file = "../analysis/SKA/ST457_dists/ST457_dist_100_SNP.clusters.tsv", delim = "\t")

clusts <- clusts %>% mutate(ST = "ST457")

clusts2 <- bind_rows(clusts2, clusts)
clusts <- read_delim(file = "../analysis/SKA/ST57_dists/ST57_dist_100_SNP.clusters.tsv", delim = "\t")

clusts <- clusts %>% mutate(ST = "ST57")

clusts2 <- bind_rows(clusts2, clusts)
clusts <- read_delim(file = "../analysis/SKA/ST648_dists/ST648_dist_100_SNP.clusters.tsv", delim = "\t")

clusts <- clusts %>% mutate(ST = "ST648")

clusts2 <- bind_rows(clusts2, clusts)

clusts <- read_delim(file = "../analysis/SKA/ST1193_dists/ST1193_dist_100_SNP.clusters.tsv", delim = "\t")

clusts <- clusts %>% mutate(ST = "ST1193")

clusts2 <- bind_rows(clusts2, clusts)

clusts <- read_delim(file = "../analysis/SKA/ST80_dists/ST80_dist_100_SNP.clusters.tsv", delim = "\t")

clusts <- clusts %>% mutate(ST = "ST80")

clusts2 <- bind_rows(clusts2, clusts)










clusts_w_metadata <- genometa %>% select(name, ST_new, Revised_Source_Niche) %>% rename("ID" = name) %>% inner_join(clusts2, by = "ID") %>% mutate(Cluster = paste0("ST", ST_new, "_", Cluster__autocolour))





simple_clusts <- clusts_w_metadata %>%
  mutate(Val_fil = 1) %>% 
  pivot_wider(names_from = Revised_Source_Niche, values_from = Val_fil, values_fill = 0, values_fn = sum) %>%
  group_by(Cluster) %>%
  summarise(Human = sum(Human), `Wild Animal` = sum(`Wild Animal`), `Companion Animal` = sum(`Companion Animal`), Food = sum(Food), Livestock = sum(Livestock), Environmental = sum(Environmental)) %>%
  ungroup() %>%
  mutate(sum_strains = Human + `Wild Animal` + `Companion Animal` + Food + Livestock + Environmental) %>%
  mutate(Human = if_else(Human > 1, 1, Human)) %>%
  mutate(`Wild Animal` = if_else(`Wild Animal` > 1, 1, `Wild Animal`)) %>%
  mutate(`Companion Animal` = if_else(`Companion Animal` > 1, 1, `Companion Animal`)) %>% 
  mutate(Food = if_else(Food > 1, 1, Food)) %>% 
  mutate(Livestock = if_else(Livestock > 1, 1, Livestock)) %>%
  mutate(Environmental = if_else(Environmental > 1, 1, Environmental)) %>%
  ungroup() %>% 
  mutate(Multi_sector = Human + `Wild Animal` + `Companion Animal` + Food + Livestock + Environmental) %>%
  ungroup() %>%
  mutate(Multi_sector = if_else(Multi_sector > 1, true = 1, false = 0)) %>%
  select(Cluster, Multi_sector, sum_strains) %>%
  mutate(ST = str_replace(Cluster, "(ST[0-9]+)_.*", "\\1")) %>%
  select(Cluster, ST, Multi_sector, sum_strains)

simple_clusts %>% filter(sum_strains > 1)  %>% group_by(ST, Multi_sector) %>% tally() %>% pivot_wider(names_from = Multi_sector, values_from = n, values_fill = 0) %>% mutate(Total = `0` + `1`) %>% rename("Single-Source" = `0`, "Multi-Source" = `1`) %>% arrange(desc(`Multi-Source`))


simple_clusts %>% filter(sum_strains > 1) %>% filter(Multi_sector == 1) %>% group_by(ST) %>% summarise(Min = min(sum_strains), Median = min(sum_strains), Max = max(sum_strains), clust_count = n()) %>% arrange(desc(clust_count))


```

