---
title: "AusTrakka_Accessory_Analysis"
author: "Max Cummins"
date: "`r Sys.Date()`"
output:
    html_document:
        theme: spacelab
        toc: true
        toc_float: true
---

```{r setup, include=FALSE}
#Read in our packages
library(tidyverse)
library(vroom)
library(ggplot2)
library(xml2)
library(ggrepel)
library(caret)
library(here)

knitr::opts_chunk$set(echo = TRUE, root.dir = here::here())

#Define our not in function
`%nin%` <- Negate(`%in%`)
```

# MobSuite Processing

Below we will explore our plasmid data as generated by MobSuite.

## Read in our tree and genomic and metadata dataset

```{r Merge_analysis_outputs, include=FALSE}

genometa <- vroom("delims/genometa_n5631.txt", show_col_types = FALSE)

#Read in our tree
tree <- ggtree::read.tree("analysis/pad.nwk")

#Remove quotes from sample names
tree$tip.label <- gsub("'", "", tree$tip.label)

#Create a list of samples in our cohort
sample_names <- tree$tip.label %>% as.data.frame() %>% rename('name' = '.')

```

## Read in our plasmid cluster data

```{r}
if(!exists("clusters")){
        if(file.exists("delims/plasmid_cluster_data.txt")){
           clusters <- vroom("delims/plasmid_cluster_data.txt")     
        }else{source("scripts/plasmid_data_aggregate.R")}
}
```



## MobTyper Results
```{r}
#Read in our mobtyper data
mobtyper_df <- vroom("analysis/mobtyper_results_concatenated.txt", show_col_types = FALSE)

#Filter in our mobtyper data and rename the sample ID column to name
mobtyper_df <- mobtyper_df %>% rename('name' = sample_id) %>% mutate(name = str_replace(name, ":.*", "")) %>% filter(name %in% tree$tip.label)
```


## Entropy attempt
```{r}
detected_clusters <- clusters %>% filter(primary_cluster_id %in% unique(mobtyper_df$primary_cluster_id))


ST_entropy<- genometa %>%
        group_by(ST_new) %>%
        add_count(name = 'Count_of_ST') %>%
        ungroup() %>%
        select(name, ST_new, Count_of_ST) %>%
        inner_join(mobtyper_df, by = "name") %>%
        group_by(ST_new, primary_cluster_id) %>%
        summarise(counts = n(), Count_of_ST) %>%
        unique() %>%
        group_by(ST_new, primary_cluster_id) %>% 
        mutate(ST_cluster_prob = counts/Count_of_ST) %>%
        mutate(log_ST_cluster_prob = log(ST_cluster_prob)) %>%
        mutate(prod_ST_cluster_prob_and_log = ST_cluster_prob*log_ST_cluster_prob) %>%
        group_by(primary_cluster_id) %>%
        mutate(entropy = abs(sum(prod_ST_cluster_prob_and_log))) %>%
        mutate('ST_entropy' = entropy) %>% select(primary_cluster_id, ST_entropy) %>% unique()

Phylogroup_entropy <- genometa %>%
        group_by(Consensus_phylogroup) %>%
        add_count(name = 'Count_of_Phylogroup') %>%
        ungroup() %>%
        select(name, Consensus_phylogroup, Count_of_Phylogroup) %>%
        inner_join(mobtyper_df, by = "name") %>%
        group_by(Consensus_phylogroup, primary_cluster_id) %>%
        summarise(counts = n(), Count_of_Phylogroup) %>%
        unique() %>%
        group_by(Consensus_phylogroup, primary_cluster_id) %>% 
        mutate(Phylogroup_cluster_prob = counts/Count_of_Phylogroup) %>%
        mutate(log_Phylogroup_cluster_prob = log(Phylogroup_cluster_prob)) %>%
        mutate(prod_Phylogroup_cluster_prob_and_log = Phylogroup_cluster_prob*log_Phylogroup_cluster_prob) %>%
        group_by(primary_cluster_id) %>%
        mutate(entropy = abs(sum(prod_Phylogroup_cluster_prob_and_log))) %>%
        mutate('Phylogroup_entropy' = entropy) %>%
        select(primary_cluster_id, Phylogroup_entropy) %>% unique()

Source_entropy<- genometa %>%
        group_by(Revised_Source_Niche) %>%
        add_count(name = 'Count_of_Revised_Source_Niche') %>%
        ungroup() %>%
        select(name, Revised_Source_Niche, Count_of_Revised_Source_Niche) %>%
        inner_join(mobtyper_df, by = "name") %>%
        group_by(Revised_Source_Niche, primary_cluster_id) %>%
        summarise(counts = n(), Count_of_Revised_Source_Niche) %>%
        unique() %>%
        group_by(Revised_Source_Niche, primary_cluster_id) %>% 
        mutate(Revised_Source_Niche_cluster_prob = counts/Count_of_Revised_Source_Niche) %>%
        mutate(log_Revised_Source_Niche_cluster_prob = log(Revised_Source_Niche_cluster_prob)) %>%
        mutate(prod_Revised_Source_Niche_cluster_prob_and_log = Revised_Source_Niche_cluster_prob*log_Revised_Source_Niche_cluster_prob) %>%
        group_by(primary_cluster_id) %>%
        mutate(entropy = abs(sum(prod_Revised_Source_Niche_cluster_prob_and_log))) %>%
        mutate('Source_entropy' = entropy) %>% select(primary_cluster_id, Source_entropy) %>% unique()

detected_clusters_w_entropy <- left_join(detected_clusters, Phylogroup_entropy, by = "primary_cluster_id", relationship = "many-to-many") %>% left_join(., ST_entropy, by = "primary_cluster_id", relationship = "many-to-many") %>% left_join(., Source_entropy, by = "primary_cluster_id", relationship = "many-to-many") %>% select(sample_id, matches("entropy"), Plasmid_CIA_status, Plasmid_count_of_AMR_genes, Plasmid_VF_count, matches("^inc"), `IncF RST`, everything()) %>%
        mutate(Plasmid_CIA_status = replace_na(Plasmid_CIA_status, 0)) %>%
        mutate(Plasmid_count_of_AMR_genes = replace_na(Plasmid_count_of_AMR_genes, 0)) %>%
        mutate(Plasmid_VF_count = replace_na(Plasmid_VF_count, 0))

plot_df <- detected_clusters_w_entropy %>%
        group_by(primary_cluster_id) %>%
        add_count(`rep_type(s)`) %>%
        #filter(primary_cluster_id %in% mobtyper_results$primary_cluster_id) %>%
        #Identify the most common replicon types within the remaning plasmids
        mutate(Major_reps = `rep_type(s)`[n == max(n)][1]) %>%
        select(-n) %>%
        add_count(`IncF RST`) %>%
        #filter(primary_cluster_id %in% mobtyper_results$primary_cluster_id) %>%
        #Identify the most common replicon types within the remaning plasmids
        mutate(Major_IncF_RST = `IncF RST`[n == max(n)][1]) %>%
        mutate(median_cluster_AMR_gene_count = median(Plasmid_count_of_AMR_genes)) %>%
        as.data.frame() %>% 
        group_by(primary_cluster_id) %>% 
        summarise(primary_cluster_id,
                  median_size = median(size),
                  Phylogroup_entropy,
                  ST_entropy,
                  Source_entropy,
                  Major_reps,
                  Major_IncF_RST,
                  median_VF_count = median(Plasmid_VF_count),
                  median_ARG_count = median(Plasmid_count_of_AMR_genes),
        ) %>%
        unique() 

plot_df <- mobtyper_df %>% group_by(primary_cluster_id) %>% add_count(name = "Total cluster members") %>% select(primary_cluster_id, `Total cluster members`) %>% inner_join(plot_df)


#plot_df %>% arrange(desc(`Total cluster members`)) %>% group_by(primary_cluster_id) %>% slice_head(n = 1) %>% ggplot() +
#        geom_point(aes(x = Phylogroup_entropy, y = median_VF_count, color = predicted_mobility)) +
#        scale_fill_manual(values = c('conjugative' = "red", 'non-conjugative' = 'blue', 'non-mobalizable' = 'green')) +
#        geom_text_repel(aes(x = Phylogroup_entropy, y = median_VF_count, label = primary_cluster_id))


```

```{r}
novel_and_plasmid_counts <- genometa %>%
        #Select columns of interest
        select(name, Revised_Source_Niche, Consensus_phylogroup) %>%
        #group by source to enable count
        group_by(Revised_Source_Niche) %>%
        #add source count
        add_count(name = "Source_Count") %>%
        #remove grouping
        ungroup() %>%
        #Bind to our mobtyper data
        full_join(mobtyper_df, by = "name") %>%
        #Group by sample
        group_by(name) %>%
        #Count the number of plasmids in a sample
        mutate(count_plasmids = n()) %>%
        #Correct where a count of zero is misidentified as a count of 1 for samples with no mobtyper data
        mutate(count_plasmids = if_else(is.na(md5) & count_plasmids == 1, 0, count_plasmids)) %>%
        #Add a column to indicate whether a particular plasmid is novel or not
        mutate(is_novel =  if_else(grepl("novel", primary_cluster_id), 1, 0)) %>%
        #Correct where a gene is indicated as having a non novel plasmid but actually has no plasmids
        mutate(is_novel = if_else(is.na(md5) & is_novel == 0, NA, is_novel)) %>%
        #Remove grouping
        ungroup()


source_wise_novel_total_plasmid_counts <- novel_and_plasmid_counts %>%
        #Group by Source and is novel to enable counts
        group_by(Revised_Source_Niche, is_novel) %>%
        #Add a count of novel plasmids within a given source
        add_count() %>%
        #Remove grouping
        ungroup() %>%
        #Isolate columns of interest
        select(Revised_Source_Niche, Source_Count, is_novel, n) %>%
        #Remove unnecessary rows
        unique()


source_wise_novel_total_plasmid_counts %>%
        #Make into wide format to produce a table showing the
        #count of novel and known primary_cluster_ids within a given source
        pivot_wider(names_from = "is_novel", values_from = n) %>%
        #Make a column showing what percentage of the plasmids in a source are novel
        mutate(source_wise_perc = scales::percent(`1`/`0`)) %>%
        #Make a column showing what percentage of the novel plasmids are from a given source
        mutate(novel_total_perc = scales::percent(`1`/732)) %>%
        #Make a column showing what percentage of the full collection is represented by a given source
        mutate(collection_total_perc = scales::percent(Source_Count/5631))  %>%
        #Make a column showing the average number of novel plasmids carried by strains within a given source
        mutate(adjusted_mean_carriage = `1`/Source_Count) %>%
        #Adjust this based on the mean number of novel plasmids observed within a given source
        mutate(num_strains_per_novel = 1/adjusted_mean_carriage) %>%
        #Make a column showing the mean number of all plasmids within a given source
        mutate(mean_total_plas = (`0`+`1`)/Source_Count) %>%
        #Make a column that penalises sources based on their mean count of total plasmids to proportionally reduce
        #their prevalence of carriage of novel plasmids
        mutate(adj_num_strains_per_novel = num_strains_per_novel/mean_total_plas) %>%
        #Select columns of interest
        select(Revised_Source_Niche,
               Source_Count,
               `0`,
               `1`,
               `NA`,
               adjusted_mean_carriage,
               mean_total_plas,
               num_strains_per_novel,
               adj_num_strains_per_novel,
               source_wise_perc,
               novel_total_perc,
               collection_total_perc) %>%
        unique()

anova_df_source_novel <- novel_and_plasmid_counts %>%
        select('name', 'Revised_Source_Niche','count_plasmids') %>%
        mutate(Revised_Source_Niche = factor(Revised_Source_Niche,
                                             levels = c("Food",
                                                        "Environmental",
                                                        "Livestock",
                                                        "Wild Animal",
                                                        "Human",
                                                        "Companion Animal"))) %>%
        unique()

set.seed(1)
subsampled_anova_df_source_novel <- anova_df_source_novel %>% group_by(Revised_Source_Niche) %>% sample_n(302)

subsampled_anova_df_source_novel %>% select(name) %>% as.data.frame() %>% write_delim("delims/302_subset_names.txt")

subsampled_anova_Source_novel <- aov(count_plasmids ~ Revised_Source_Niche, data = subsampled_anova_df_source_novel)

summary(subsampled_anova_Source_novel)

tukey_result <- TukeyHSD(subsampled_anova_Source_novel)



#model <- lm(n_novel ~ Revised_Source_Niche + count_plasmids, data = subsampled_anova_df_source_novel)

#summary(model)

#test_set <- anova_df_source_novel %>% filter(name %nin% subsampled_anova_df_source_novel$name)

#test_set$pred <- predict(object = model, test_set, type = "response")


#Source Colours
Source_cols <- c("Companion Animal" = "#59398d",
         "Environmental" = "#709b46",
         "Food" = "#c09f3d",
         "Human" = "#48c595",
         "Livestock" =  "#c26bbc",
         "Wild Animal" = "#b9553d")

#source_Novel_count_plot <- subsampled_anova_df_source_novel %>%
#        ggplot() +
#        geom_boxplot(aes(x = Revised_Source_Niche, y = n_novel, fill = Revised_Source_Niche)) +
#        scale_fill_manual(values = Source_cols)



```

## Table 1

```{r}
genometa %>% group_by(Consensus_phylogroup, Revised_Source_Niche) %>% tally() %>% pivot_wider(values_from = n, names_from = Revised_Source_Niche, values_fill = 0) %>% mutate(Total = `Companion Animal` + Environmental + Food + Human + Livestock + `Wild Animal`)
```
## Histogram of ST frequency
```{r}
ST_names <- genometa %>% mutate(ST_other = fct_lump_n(ST_new, 20)) %>% filter(ST_other != "Other") %>% group_by(ST_other) %>% tally() %>% filter(n > 10) %>% pull(ST_other)
ST_cols <- hues::iwanthue(length(ST_names))
names(ST_cols) <- ST_names


genometa %>% mutate(ST_other = fct_lump_n(ST_new, 20)) %>% filter(ST_other != "Other") %>% group_by(ST_other) %>% tally() %>% filter(n > 10) %>%
  ggplot() +
  geom_bar(aes(x = ST_other, y = n, fill = ST_other)) +
  scale_fill_manual(values = ST_cols)
```

```{r}
novel_and_plasmid_counts <- genometa %>%
        #Select columns of interest
        select(name, Consensus_phylogroup, Consensus_phylogroup) %>%
        #group by Phylogroup to enable count
        group_by(Consensus_phylogroup) %>%
        #add Phylogroup count
        add_count(name = "Phylogroup_Count") %>%
        #remove grouping
        ungroup() %>%
        #Bind to our mobtyper data
        full_join(mobtyper_df, by = "name") %>%
        #Group by sample
        group_by(name) %>%
        #Count the number of plasmids in a sample
        mutate(count_plasmids = n()) %>%
        #Correct where a count of zero is misidentified as a count of 1 for samples with no mobtyper data
        mutate(count_plasmids = if_else(is.na(md5) & count_plasmids == 1, 0, count_plasmids)) %>%
        #Add a column to indicate whether a particular plasmid is novel or not
        mutate(is_novel =  if_else(grepl("novel", primary_cluster_id), 1, 0)) %>%
        #Correct where a gene is indicated as having a non novel plasmid but actually has no plasmids
        mutate(is_novel = if_else(is.na(md5) & is_novel == 0, NA, is_novel)) %>%
        #Remove grouping
        ungroup()


Phylogroup_wise_novel_total_plasmid_counts <- novel_and_plasmid_counts %>%
        #Group by Phylogroup and is novel to enable counts
        group_by(Consensus_phylogroup, is_novel) %>%
        #Add a count of novel plasmids within a given Phylogroup
        add_count() %>%
        #Remove grouping
        ungroup() %>%
        #Isolate columns of interest
        select(Consensus_phylogroup, Phylogroup_Count, is_novel, n) %>%
        #Remove unnecessary rows
        unique()


Phylogroup_wise_novel_total_plasmid_counts %>%
        #Make into wide format to produce a table showing the
        #count of novel and known primary_cluster_ids within a given Phylogroup
        pivot_wider(names_from = "is_novel", values_from = n) %>%
        #Make a column showing what percentage of the plasmids in a Phylogroup are novel
        mutate(Phylogroup_wise_perc = scales::percent(`1`/`0`)) %>%
        #Make a column showing what percentage of the novel plasmids are from a given Phylogroup
        mutate(novel_total_perc = scales::percent(`1`/732)) %>%
        #Make a column showing what percentage of the full collection is represented by a given Phylogroup
        mutate(collection_total_perc = scales::percent(Phylogroup_Count/5631))  %>%
        #Make a column showing the average number of novel plasmids carried by strains within a given Phylogroup
        mutate(adjusted_mean_carriage = `1`/Phylogroup_Count) %>%
        #Adjust this based on the mean number of novel plasmids observed within a given Phylogroup
        mutate(num_strains_per_novel = 1/adjusted_mean_carriage) %>%
        #Make a column showing the mean number of all plasmids within a given Phylogroup
        mutate(mean_total_plas = (`0`+`1`)/Phylogroup_Count) %>%
        #Make a column that penalises Phylogroups based on their mean count of total plasmids to proportionally reduce
        #their prevalence of carriage of novel plasmids
        mutate(adj_num_strains_per_novel = num_strains_per_novel/mean_total_plas) %>%
        #Select columns of interest
        select(Consensus_phylogroup,
               Phylogroup_Count,
               `0`,
               `1`,
               `NA`,
               adjusted_mean_carriage,
               mean_total_plas,
               num_strains_per_novel,
               adj_num_strains_per_novel,
               Phylogroup_wise_perc,
               novel_total_perc,
               collection_total_perc) %>%
        unique()

anova_df_Phylogroup_novel <- novel_and_plasmid_counts %>%
        select('name', 'Consensus_phylogroup','count_plasmids') %>%
        #mutate(Consensus_phylogroup = factor(Consensus_phylogroup,
        #                                     levels = c("Food",
        #                                                "Environmental",
        #                                                "Livestock",
        #                                                "Wild Animal",
        #                                                "Human",
        #                                                "Companion Animal"))) %>%
        unique()
```

```{r}
set.seed(1)
subsampled_anova_df_Phylogroup_novel <- anova_df_Phylogroup_novel %>% filter(Consensus_phylogroup %in% c('A', 'B1', 'B2', 'C', 'D', 'E', 'F', 'G')) %>% group_by(Consensus_phylogroup) %>% sample_n(123)

subsampled_anova_Phylogroup_novel <- aov(count_plasmids ~ Consensus_phylogroup, data = subsampled_anova_df_Phylogroup_novel)

summary(subsampled_anova_Phylogroup_novel)

tukey_result <- TukeyHSD(subsampled_anova_Phylogroup_novel)

tukey_vis_df <- tukey_result$Consensus_phylogroup %>% as.data.frame() %>% rownames_to_column("phylogroup1") %>% mutate(phylogroup2 = str_replace(phylogroup1, ".*-", ""), phylogroup1 = str_replace(phylogroup1, "-.*", "")) %>% filter(`p adj` < 0.05) %>% select(matches("phylogroup"), diff) 

tukey_vis_df %>% select(phylogroup2, phylogroup1, everything()) %>% mutate(diff = -diff) %>% rename("tmp" = phylogroup1) %>% rename("phylogroup1" = phylogroup2) %>% rename("phylogroup2" = tmp) %>% bind_rows(tukey_vis_df) %>% pivot_wider(names_from = "phylogroup2", values_from = "diff", values_fill = 0) %>% column_to_rownames('phylogroup1') %>% select(all_of(rownames(.))) %>% pheatmap::pheatmap(display_numbers = T)


#model <- lm(n_novel ~ Consensus_phylogroup + count_plasmids, data = subsampled_anova_df_Phylogroup_novel)

#summary(model)

#test_set <- anova_df_Phylogroup_novel %>% filter(name %nin% subsampled_anova_df_Phylogroup_novel$name)

#test_set$pred <- predict(object = model, test_set, type = "response")


#Phylogroup Colours
Phylogroup_cols <- c("Companion Animal" = "#59398d",
         "Environmental" = "#709b46",
         "Food" = "#c09f3d",
         "Human" = "#48c595",
         "Livestock" =  "#c26bbc",
         "Wild Animal" = "#b9553d")

#Phylogroup_Novel_count_plot <- subsampled_anova_df_Phylogroup_novel %>%
#        ggplot() +
#        geom_boxplot(aes(x = Consensus_phylogroup, y = n_novel, fill = Consensus_phylogroup)) +
#        scale_fill_manual(values = Phylogroup_cols)



```

```{r}
POCs <- clusters %>% group_by(primary_cluster_id) %>%
        mutate(across(matches("Plasmid_"), ~ replace_na(.x, 0))) %>% 
        mutate(POC_vir_status = if_else(median(Plasmid_VF_count) > 1, 1, 0)) %>%
        mutate(POC_cia_status = if_else(mean(Plasmid_CIA_status) > 0.75, 1, 0)) %>%
        mutate(POC_res_status = if_else(median(Plasmid_count_of_AMR_genes) > 0, 1, 0)) %>%
        mutate(POC_more_res_status = if_else(median(Plasmid_count_of_AMR_genes) >= 3, 1, 0)) %>%
        select(primary_cluster_id, POC_res_status, POC_cia_status, POC_more_res_status, POC_vir_status) %>% unique()
        

genometa_mobtyper <- genometa %>% select(name, Revised_Source_Niche, Consensus_phylogroup, ST_new, Strain_CIA_status, Strain_VF_count, Strain_count_of_AMR_genes) %>% full_join(mobtyper_df, by = "name") 


has_res_vir_cia_POC <- genometa_mobtyper %>%
        left_join(POCs, by = "primary_cluster_id") %>%
        mutate(across(matches("POC"), ~ replace_na(.x, 0))) %>% 
        group_by(name) %>%
        summarise(name,
                  Revised_Source_Niche,
                  Consensus_phylogroup,
                  ST_new,
                  Strain_CIA_status,
                  Strain_VF_count,
                  Strain_count_of_AMR_genes,
                  has_res_POC = sum(POC_res_status),
                  has_cia_POC = sum(POC_cia_status),
                  has_vir_POC = sum(POC_vir_status)) %>%
        mutate(has_res_POC = if_else(has_res_POC >= 1, 1, 0),
               has_cia_POC = if_else(has_cia_POC >= 1, 1, 0),
               has_vir_POC = if_else(has_vir_POC >= 1, 1, 0)) %>%
        unique()

has_res_vir_cia_POC %>% ggplot() + geom_boxplot(aes(x = as.character(has_res_POC), y = Strain_count_of_AMR_genes))
has_res_vir_cia_POC %>% ggplot() + geom_boxplot(aes(x = as.character(has_vir_POC), y = Strain_VF_count))
has_res_vir_cia_POC %>% group_by(has_cia_POC) %>% summarise(mean(Strain_CIA_status))

t.test(has_res_vir_cia_POC[has_res_vir_cia_POC$has_vir_POC == 0,"Strain_VF_count"], has_res_vir_cia_POC[has_res_vir_cia_POC$has_vir_POC == 1,"Strain_VF_count"])

#Read in our libraries
library(highcharter)
library(htmlwidgets)
library(tidyverse)


## Meat of the sankey diagram vis
        #Select columns you want to visualise
sankey1 <- has_res_vir_cia_POC %>%
        as.data.frame() %>%
        select(has_cia_POC, Strain_CIA_status) %>%
        mutate(Strain_CIA_status = case_when(Strain_CIA_status == 0 ~ "CIA-Sus",
                                             Strain_CIA_status == 1 ~ "CIA-Res")) %>%
        mutate(has_cia_POC = case_when(has_cia_POC == 0 ~ as.factor("no CIA POC"),
                                             has_cia_POC == 1 ~ as.factor("CIA POC")))

#Add a count of a given variable within a given column 
sankey1 <- apply(sankey1,
                 MARGIN = 2,
                 FUN = function(x) paste0(x,
                                          " (",
                                          as.data.frame(table(x)[x])[,2],
                                          "/", as.character(nrow(has_res_vir_cia_POC)),
                                          ")"))

#Convert back to dataframe
sankey1 <- as.data.frame(sankey1)

#Visualisation step
hchart(data_to_sankey(sankey1), "sankey", name = "Collection Breakdown")  %>%
        hc_plotOptions(series = list(dataLabels = list(style = list(fontSize = "20px"))))
```

```{r}
POCs_2 <- clusters %>% group_by(primary_cluster_id) %>%
        mutate(across(matches("Plasmid_"), ~ replace_na(.x, 0))) %>% 
        mutate(POC_vir_status = if_else(median(Plasmid_VF_count) > 1, 1, 0)) %>%
        mutate(median_vir_count = median(Plasmid_VF_count)) %>%
        mutate(POC_cia_status = if_else(mean(Plasmid_CIA_status) > 0.75, 1, 0)) %>%
        mutate(POC_res_status = if_else(median(Plasmid_count_of_AMR_genes) > 0, 1, 0)) %>%
        mutate(POC_more_res_status = if_else(median(Plasmid_count_of_AMR_genes) >= 3, 1, 0)) %>%
        mutate(median_arg_count = median(Plasmid_count_of_AMR_genes)) %>%
        add_count(`rep_type(s)`) %>%
        mutate(Major_reps = `rep_type(s)`[n == max(n)][1]) %>%
        select(primary_cluster_id, Major_reps, `rep_type(s)`, POC_res_status, POC_cia_status, POC_more_res_status, POC_vir_status, median_vir_count, median_arg_count, n) %>% unique() %>%
        mutate(`Plasmid Representatives` = sum(n))

POCs_2 <- mobtyper_df %>% filter(name %in% tree$tip.label) %>% group_by(primary_cluster_id) %>% summarise(`Cluster count` = n()) %>% inner_join(POCs_2, by = 'primary_cluster_id')

POCs_2 <- POCs_2 %>%
        select(-n, -`rep_type(s)`) %>%
        unique() %>%
        mutate(simple_rep = str_replace(Major_reps, ",.*", "")) %>%
        select(primary_cluster_id, Major_reps, simple_rep, everything()) %>%
        mutate(simple_rep = case_when(str_detect(simple_rep, regex("^rep_", multiline = TRUE)) ~ "Other",
                                      str_detect(simple_rep, regex("^IncF.*", multiline = TRUE)) ~ "IncF",
                                      str_detect(simple_rep, regex("^IncH.*", multiline = TRUE)) ~ "IncH",
                                      str_detect(simple_rep, regex("^IncQ.*", multiline = TRUE)) ~ "IncQ",
                                      str_detect(simple_rep, regex("^IncF.*", multiline = TRUE)) ~ "IncF",
                                      str_detect(simple_rep, regex("^Col.*", multiline = TRUE)) ~ "Col",
                                      str_detect(simple_rep, regex("^IncX.*", multiline = TRUE)) ~ "IncX",
                                      str_detect(simple_rep, regex("^IncI.*", multiline = TRUE)) ~ "IncI",
                                      str_detect(simple_rep, regex("^Inc[0-9]+", multiline = TRUE)) ~ "Inc (Other)",
                                      is.na(simple_rep) ~ "Unknown",
                                      str_detect(simple_rep, regex("^-$", multiline = TRUE)) ~ "Unknown",
                                      TRUE ~ simple_rep))

POCs_2 <- POCs_2 %>% inner_join(Phylogroup_entropy, by = "primary_cluster_id") %>% unique()

rep_cols <- c('#e6194b', '#3cb44b', '#ffe119', '#4363d8', '#f58231', '#911eb4', '#46f0f0', '#f032e6', '#bcf60c', '#fabebe', '#008080', '#e6beff', '#9a6324', '#800000', '#aaffc3', '#808000', '#ffd8b1', '#000075', '#808080', '#ffffff', '#000000')

names(rep_cols) <- unique(sort(POCs_2$simple_rep))

normaliser <- function(x){(x-min(x))/(max(x)-min(x))}

POCs_2$normalised_phylogroup_entropy <- normaliser(POCs_2$Phylogroup_entropy)

POCs_2 %>% ggplot() +
        geom_point(aes(x = `normalised_phylogroup_entropy`, y = `Cluster count`, color = simple_rep)) + geom_text_repel(aes(x = normalised_phylogroup_entropy, y = `Cluster count`, label = primary_cluster_id)) + 
        scale_color_manual(values = rep_cols)


POCs_2 %>% filter(median_vir_count > 0) %>% rename('Phylogenetic promiscuity (normalised)' = normalised_phylogroup_entropy, 'Frequency' = `Cluster count`, 'Replicon Type' = simple_rep, 'Median VAG Count' = median_vir_count) %>%
        ggplot()  +
        geom_point(aes(x = `Phylogenetic promiscuity (normalised)`, y = `Frequency`, color = `Replicon Type`, size = `Median VAG Count`)) +
        geom_text_repel(aes(x = `Phylogenetic promiscuity (normalised)`, y = `Frequency`, label = primary_cluster_id)) + 
        scale_color_manual(values = rep_cols) +
        theme(axis.text.x = element_text(size = 16),
              axis.text.y = element_text(size = 16),
              axis.title.x = element_text(size = 16),
              axis.title.y = element_text(size = 16),
              legend.text=element_text(size=12))

POCs_2 %>% filter(median_arg_count > 0) %>% ggplot()  +
        geom_point(aes(x = `normalised_phylogroup_entropy`, y = `Cluster count`, color = simple_rep, size = `median_arg_count`, shape = as.factor(`POC_cia_status`))) + geom_text_repel(aes(x = `normalised_phylogroup_entropy`, y = `Cluster count`, label = primary_cluster_id)) + 
        scale_color_manual(values = rep_cols)

POCs_2 %>% ggplot()  +
        geom_point(aes(x = `normalised_phylogroup_entropy`, y = `Cluster count`, color = simple_rep, shape = as.factor(`POC_cia_status`))) + geom_text_repel(aes(x = `normalised_phylogroup_entropy`, y = `Cluster count`, label = primary_cluster_id)) + 
        scale_color_manual(values = rep_cols)
```


```{r}
genometa_vf_count <- genometa %>% select(name, matches("vfdb")) %>% column_to_rownames("name") %>% mutate(Strain_VF_count = rowSums(.)) %>% rownames_to_column("name") %>% select(name, Strain_VF_count)

genometa<- genometa_vf_count %>% left_join(genometa, by = 'name')

strain_ARG_VFs <- genometa %>% select(name, Strain_count_of_AMR_genes, Strain_VF_count)

strain_ARG_VFs_w_clusters<- strain_ARG_VFs %>% full_join(mobtyper_df, by = "name") %>% select(name, Strain_count_of_AMR_genes, Strain_VF_count, primary_cluster_id, secondary_cluster_id)

cluster_mean_vf_arg <- clusters %>% filter(`IncF RST` != "F-:A-:B-") %>% select(primary_cluster_id, Plasmid_count_of_AMR_genes, Plasmid_VF_count) %>% group_by(primary_cluster_id) %>% summarise(mean_ref_cluster_amr_count = mean(Plasmid_count_of_AMR_genes), mean_ref_cluster_vf_count = mean(Plasmid_VF_count))

cluster_mean_vf_arg <- cluster_mean_vf_arg %>% mutate(mean_ref_cluster_amr_count = replace_na(mean_ref_cluster_amr_count, 0)) %>% mutate(mean_ref_cluster_vf_count = replace_na(mean_ref_cluster_vf_count, 0))

ref_vs_strain_amr_vir <- strain_ARG_VFs_w_clusters %>% inner_join(cluster_mean_vf_arg, by = 'primary_cluster_id') %>% as.data.frame()

ref_vs_strain_amr_vir <- ref_vs_strain_amr_vir %>% filter(!is.na(primary_cluster_id))

test_linear_model <- lm(Strain_count_of_AMR_genes ~ primary_cluster_id + mean_ref_cluster_amr_count, data = ref_vs_strain_amr_vir)

summary(test_linear_model)


#Read in abricate data for tea dataset
abricateR::abricateR(abricate_in = "/Users/131785/Dropbox/PostDoc/Austrakka/AusTrakka_R/analysis/TEA/genotype.txt", output = "TEA_abricate", writecsv = FALSE)


mobtyper_contig_wise <- read_delim("analysis/TEA/mobtyper_results.txt", 
    delim = "\t", escape_double = FALSE, 
    trim_ws = TRUE)


#Read in our mobtyper data
mobtyper_contig_wise <- mobtyper_contig_wise %>% rename('name' = sample_id) %>% filter(name %in% tree$tip.label)

small_mobtyper_hits <- mobtyper_contig_wise %>% select(name, contig_id, primary_cluster_id) %>% mutate(contig_id = str_replace(contig_id, " .*", ""))

co_occurence_df <- TEA_abricate_co_occurence_N90L90 %>% rename('contig_id' = SEQUENCE) %>% left_join(small_mobtyper_hits, by = c("name", "contig_id"))

co_occurence_df_wide <- co_occurence_df %>%
        group_by(name, primary_cluster_id) %>%
        summarise(same_scaff = paste0(same_scaff, collapse = " ")) %>%
        #filter(primary_cluster_id %in% decat_known_PPCs_known_rep_wide$primary_cluster_id, primary_cluster_id != is.na(primary_cluster_id)) %>%
        mutate(same_scaff2 = strsplit(same_scaff, " ")) %>%
     unnest(same_scaff2) %>% 
     arrange(same_scaff2) %>%  
     pivot_wider(names_from = same_scaff2,
            names_repair = "universal", values_from = same_scaff2, 
        values_fill = 0, values_fn = length)

contig_wise_vf_counts <- co_occurence_df_wide %>% filter(primary_cluster_id %in% ref_vs_strain_amr_vir$primary_cluster_id) %>% mutate(rownem = paste(name, primary_cluster_id, sep = ",")) %>% column_to_rownames("rownem") %>% select(matches('vfdb')) %>% mutate(plasmid_contig_vf_count = rowSums(.)) %>% select(plasmid_contig_vf_count) %>% rownames_to_column("name") %>% mutate(primary_cluster_id = str_replace(name, ".*,", "")) %>% mutate(name = str_replace(name, ",.*", "")) %>% select(name, primary_cluster_id, plasmid_contig_vf_count)

ref_vs_strain_amr_vir <- left_join(ref_vs_strain_amr_vir, contig_wise_vf_counts, by = c("name", "primary_cluster_id"))

test_linear_model <- lm(mean_ref_cluster_vf_count ~ plasmid_contig_vf_count, data = ref_vs_strain_amr_vir)

summary(test_linear_model)

#specify the cross-validation method
ctrl <- trainControl(method = "cv", number = 5)

ref_vs_strain_amr_vir <- ref_vs_strain_amr_vir %>% mutate(plasmid_contig_vf_count = replace_na(plasmid_contig_vf_count, 0))

#fit a regression model and use LOOCV to evaluate performance
model1 <- train(mean_ref_cluster_vf_count ~ plasmid_contig_vf_count, data = ref_vs_strain_amr_vir, method = "lm", trControl = ctrl)




contig_wise_amr_counts <- co_occurence_df_wide %>% filter(primary_cluster_id %in% ref_vs_strain_amr_vir$primary_cluster_id) %>% mutate(rownem = paste(name, primary_cluster_id, sep = ",")) %>% column_to_rownames("rownem") %>% select(matches('card')) %>% mutate(plasmid_contig_amr_count = rowSums(.)) %>% select(plasmid_contig_amr_count) %>% rownames_to_column("name") %>% mutate(primary_cluster_id = str_replace(name, ".*,", "")) %>% mutate(name = str_replace(name, ",.*", "")) %>% select(name, primary_cluster_id, plasmid_contig_amr_count)

ref_vs_strain_amr_vir <- left_join(ref_vs_strain_amr_vir, contig_wise_amr_counts, by = c("name", "primary_cluster_id"))

test_linear_model <- lm(plasmid_contig_amr_count.x ~ primary_cluster_id, data = ref_vs_strain_amr_vir)

summary(test_linear_model)
```


Lets start by making a table which shows the prevalence of 
```{r}
#Define the columns we wish to split (taking everything other than name)
#To do this we identify any columns containing a comma
mob_cols_for_split <-
        mobtyper_df %>%
        dplyr::select(`rep_type(s)`) %>%
        colnames()

#Split concatenated data (comma separated into extra rows)
decat_mobtyper_df <- 
        splitstackshape::cSplit(
                indt = mobtyper_df,
                splitCols = mob_cols_for_split,
                sep = ",",
                direction = "long")

#Create a table that shows the carriage of the sum of a given replicon type by a strain
decat_mobtyper_replicons_wide <- decat_mobtyper_df %>%
        dplyr::select(name, `rep_type(s)`) %>%
        mutate(val = 1) %>%
        group_by(name, `rep_type(s)`) %>%
        summarise(counts = n(), .groups = "drop") %>%
        pivot_wider(id_cols = name, names_from = `rep_type(s)`, values_from = counts, values_fill = 0) %>%
        right_join(sample_names, by = "name") %>%
        arrange(match(name, sample_names$name)) %>% 
        dplyr::select(-`-`) %>%
        column_to_rownames("name") %>%
        dplyr::select(sort(colnames(.))) %>%
        mutate(across(everything(), ~ replace_na(.x, 0)))
```

## NMDS

Lets now take a subset of our dataset and see if their plasmid replicon profiles
cluster by Source, ST or Phylogroup.
```{r MDS, eval=FALSE, include=FALSE}
#Read in necesary packages
library(MASS)
library(vegan)
library(ellipse)
library(grid)
library(gridExtra)

#Set seed for reproducibility
set.seed(1)

#Create a list of sample indices to subsample our data with
        #We will take approximately one quarter of the data (`round(nrow(sample_names)/4)`)
sample_name_subset_idx <- sample(1:nrow(sample_names), round(nrow(sample_names)/4))

#Create our subset
plas_df <- decat_mobtyper_replicons_wide[sample_name_subset_idx,]

#Change data to be binary (Note this removes instances where samples have multiple counts of a given replicon)
plas_df[plas_df > 1] <- 1

#Remove rows which only contain 0s
plas_df_MDS <- plas_df %>% filter(rowSums(.) != 0) 

#Remove duplicate rows
plas_df_MDS <- unique(plas_df_MDS)

#create distance object 
plasdist <- vegdist(wisconsin(plas_df_MDS), method = "jaccard")
# Get the baseline solution: start with cmdscale
mds.null.plas <- isoMDS(plasdist, tol=1e-7)
## See if you can get any better.
  repeat{
    mds.1.plas <- isoMDS(plasdist, k=2, initMDS(plasdist, k = 2), maxit=200, trace=FALSE, tol=1e-7)
    print(mds.1.plas$stress)
    print(mds.null.plas$stress)
    if(mds.1.plas$stress < mds.null.plas$stress) break
  }

#read data into variable data
mds.data <- genometa[sample_name_subset_idx,]
rownames(mds.data) <- genometa %>% pull(name) %>% .[sample_name_subset_idx]
mds.metadata <- mds.data %>% dplyr::dplyr::select(name, Revised_Source_Niche, `IncF_RST..IncF RST`, clermontyping..clermontyping_phylogroup, ST_new)

#Create a column for ST_other
mds.metadata <- mds.metadata %>% mutate(ST_other = fct_lump_n(ST_new, 19))

# Scale solutions ("fix translation, rotation and scale")
mds.null.plas <- postMDS(mds.null.plas, plasdist)
mds.1.plas <- postMDS(mds.1.plas, plasdist)
# Compare solutions
#plot(procrustes(mds.1.plas, mds.null.plas))

#convert to dataframe
mds.null.plas.df <- as.data.frame(mds.null.plas,
                             attr(x = mds.null.plas,
                                which = "dimnames"))
mds.1.plas.df <- as.data.frame(mds.1.plas, 
                          attr(x = mds.1.plas, 
                              which = "dimnames"))

mds.1.plas.df$name <- mds.1.plas$points %>% rownames() 
mds.1.plas.df <- left_join(mds.1.plas.df, mds.metadata)

mds.1.plas.df$ST_new <- as.factor(mds.1.plas.df$ST_new)

#ST Other colours
ST_other_cols <- c('#e6194b', '#3cb44b', '#ffe119', '#4363d8', '#f58231', '#911eb4', '#46f0f0', '#f032e6', '#bcf60c', '#fabebe', '#008080', '#e6beff', '#9a6324', '#fffac8', '#800000', '#aaffc3', '#808000', '#ffd8b1', '#000075', '#808080', 'grey50', '#000000')

names(ST_other_cols) <- sort(unique(mds.metadata$ST_other))

#Source Colours
Source_cols <- c("Companion Animal" = "#59398d",
         "Environmental" = "#709b46",
         "Food" = "#c09f3d",
         "Human" = "#48c595",
         "Livestock" =  "#c26bbc",
         "Wild Animal" = "#b9553d")


#Define our colours for our phylogroups
clermont_cols <- c('#ffe119', '#4363d8', '#f58231', '#dcbeff', '#800000', '#000075', '#9A6324', 'red', '#3cb44b',  'white', '#a9a9a9', 'black')

#Assign our color names for phylogroups
names(clermont_cols) <- c("A", "B1",  "B2", "C", "D", "E", "F", "G", "E or cladeI", "cladeI", "cladeIII", "cladeIV")



#Plasmid by Phylogroup
a1 <- ggplot() +
  geom_point(data=mds.1.plas.df,
             aes(x=points.1,
                 y=points.2,
                 color = clermontyping..clermontyping_phylogroup), 
             size = 1.5) + # add the point markers
  stat_ellipse(aes(x=mds.1.plas.df$points.1,
                   y=mds.1.plas.df$points.2, 
                   color = mds.1.plas.df$clermontyping..clermontyping_phylogroup)) +
  scale_color_manual(breaks = names(clermont_cols), values = clermont_cols)+
  ggtitle("Nonmetric MDS - Plasmid Profile by Phylogroup")+
  xlab("Dimension 1")+
  ylab("Dimension 2")+
  coord_equal(xlim = c(-3, 4), ylim = c(-1.5, 2)) +
  theme(plot.title = element_text(hjust = 0.5),
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())

#Plasmid by Source
b1 <- ggplot() +
  geom_point(data=mds.1.plas.df,
             aes(x=points.1,
                 y=points.2,
                 color = ST_other), 
             size = 1.5) + # add the point markers
  stat_ellipse(aes(x=mds.1.plas.df$points.1,
                   y=mds.1.plas.df$points.2, 
                   color = mds.1.plas.df$ST_other)) +
  scale_color_manual(breaks = names(Source_cols), values = Source_cols)+
  ggtitle("Nonmetric MDS - Plasmid Profile by Source")+
  xlab("Dimension 1")+
  ylab("Dimension 2")+
  coord_equal(xlim = c(-3, 4), ylim = c(-1.5, 2)) +
  theme(plot.title = element_text(hjust = 0.5),
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())

#Plasmid by Source
c1 <- ggplot() +
  geom_point(data=mds.1.plas.df,
             aes(x=points.1,
                 y=points.2,
                 color = Revised_Source_Niche), 
             size = 1.5) + # add the point markers
  stat_ellipse(aes(x=mds.1.plas.df$points.1,
                   y=mds.1.plas.df$points.2, 
                   color = mds.1.plas.df$ST_other)) +
  scale_color_manual(breaks = names(ST_other_cols), values = ST_other_cols)+
  ggtitle("Nonmetric MDS - Plasmid Profile by ST")+
  xlab("Dimension 1")+
  ylab("Dimension 2")+
  coord_equal(xlim = c(-3, 4), ylim = c(-1.5, 2)) +
  theme(plot.title = element_text(hjust = 0.5),
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())

detach("package:vegan", unload=TRUE)
```

```{r}

#Read in our mobtyper data
mobtyper_df <- vroom("analysis/TEA/mobtyper_results_concatenated.txt", show_col_types = FALSE)

#Filter in our mobtyper data and rename the sample ID column to name
mobtyper_df <- mobtyper_df %>% rename('name' = sample_id) %>% mutate(name = str_replace(name, ":.*", "")) %>% filter(name %in% tree$tip.label)

#Make a table of our genomes IncF RSTs
IncF_RSTs <- genometa %>% select(name, `IncF_RST..IncF RST`) %>% rename("IncF RST" = `IncF_RST..IncF RST`)

#Bind our IncF RST data to our mobtyper data for our genomes
mobtyper_df <- left_join(IncF_RSTs, mobtyper_df, by = "name") %>%
        filter(name %in% tree$tip.label) %>%
        mutate(paste_RST_cluster = paste(primary_cluster_id, `IncF RST`))

#Remove anything that has a negative IncF type (i.e. no F repl)
detected_clusters <- clusters %>%
        filter(primary_cluster_id %in% mobtyper_df$primary_cluster_id)

known_PPCs_known_rep <- mobtyper_df %>%
        group_by(primary_cluster_id) %>%
        tally(sort = T) %>%
        left_join(detected_clusters) %>% 
        filter(!is.na(primary_cluster_id)) %>%
        select(primary_cluster_id, n, size, `rep_type(s)`) %>%
        group_by(primary_cluster_id) %>%
    mutate(median_size = median(size)) %>%
        select(-size) %>% filter(`rep_type(s)` != "-", !grepl("novel", primary_cluster_id))

known_PPCs_known_rep_cols_for_split <- known_PPCs_known_rep %>% as.data.frame() %>% select(`rep_type(s)`) %>% colnames()

#Split concatenated data (comma separated into extra rows)
decat_known_PPCs_known_rep<- 
        splitstackshape::cSplit(
                indt = known_PPCs_known_rep,
                splitCols = known_PPCs_known_rep_cols_for_split,
                sep = ",",
                direction = "long")

#Create a table that shows the carriage of the sum of a given replicon type by a strain
decat_known_PPCs_known_rep_wide <- decat_known_PPCs_known_rep %>%
    mutate(val = 1) %>%
    filter(!is.na(primary_cluster_id)) %>%
        group_by(primary_cluster_id) %>%
        pivot_wider(id_cols = c(primary_cluster_id, n, median_size), names_from = `rep_type(s)`, values_from = val, values_fn = sum, values_fill = 0) 

decat_known_PPCs_known_rep_wide <- detected_clusters %>% group_by(primary_cluster_id) %>% tally() %>% rename("count_PPC_in_DB" = n) %>% full_join(decat_known_PPCs_known_rep_wide)

counts_w_replicons_hit <- genometa %>% select(name, Revised_Source_Niche, clermontyping..clermontyping_phylogroup) %>% inner_join(mobtyper_df) %>% group_by(primary_cluster_id, Revised_Source_Niche) %>% tally() %>% pivot_wider(names_from = Revised_Source_Niche, values_from = n, values_fill = 0) %>% full_join(decat_known_PPCs_known_rep_wide) %>% select(1:10, all_of(sort(colnames(.)[11:ncol(.)]))) %>% arrange(desc(n)) %>% as.data.frame() %>% mutate(across(c(11:118), .fns = ~ ./count_PPC_in_DB))

detected_cluster_by_incs<- clusters %>% select(primary_cluster_id, matches("^inc"), size) %>% filter(primary_cluster_id %in% counts_w_replicons_hit$primary_cluster_id)


#Query internal genomes clusters
mobtyper_df %>% filter(primary_cluster_id == "AA176") %>% select(name, `IncF RST`) %>% left_join(genometa) %>% group_by(Revised_Source_Niche) %>% tally(sort = T) %>% mutate(perc = scales::percent(n/435)) %>% View()

#Query db plasmid clusters
clusters %>% filter(primary_cluster_id == "AA337") %>% select(matches("card")) %>% select(where(~ is.numeric(.x) && sum(.x) !=0 )) %>% colSums() %>% sort()
```

```{r}
#Read in our cgMLST allelic distances
allelic_distances <- vroom(file = "analysis/TEA/cgmlst_coreugate/pad.txt", show_col_types = FALSE)

#Rename the name column
allelic_distances <- allelic_distances %>% rename(name = '5631')

#Change to long format
cgMLST_dists <- pivot_longer(allelic_distances, cols = 2:5632, names_repair = 'minimal')

#Rename name columns
colnames(cgMLST_dists) <- c('Sample 1', 'Sample 2', 'value')

if(!exists("genometa")){
        genometa <- vroom("delims/genometa_n5631.txt", show_col_types = FALSE)
}

#Create a dataframe we can bind to the distances
cgMLST_binder <- genometa %>% select(name, ColV, Revised_Source_Niche, Revised_Source_Type, Strain_MDR_status, Strain_CIA_status, clermontyping..clermontyping_phylogroup, Consensus_phylogroup, ST_new)

cgMLST_binder1 <- cgMLST_binder %>% rename(`Sample 1` = name)
cgMLST_binder2 <- cgMLST_binder %>% rename(`Sample 2` = name)

rm(cgMLST_binder)

cgMLST_dists2 <- cgMLST_dists %>% full_join(cgMLST_binder1, by = "Sample 1")

rm(cgMLST_dists)

cgMLST_dists2 <- cgMLST_dists2 %>% full_join(cgMLST_binder2, by = "Sample 2")

#Reorder columns (Sorry for the gibberish...)
cgMLST_dists2 <- cgMLST_dists2 %>% select(all_of(paste0(rev(colnames(cgMLST_binder1)[2:ncol(cgMLST_binder1)]),".x")), `Sample 1`, value, `Sample 2`, matches("\\.y"))

#Read in our mobtyper data if it doesnt exist
if(!exists("mobtyper_df")){
        mobtyper_df <- vroom("analysis/TEA/mobtyper_results_concatenated.txt", show_col_types = FALSE)

        #Filter in our mobtyper data and rename the sample ID column to name
        mobtyper_df <- mobtyper_df %>%
                rename('name' = sample_id) %>%
                mutate(name = str_replace(name, ":.*", "")) %>%
                filter(name %in% allelic_distances$name)
        }

PPC_by_count <- mobtyper_df %>% group_by(primary_cluster_id) %>% tally(sort = T)

number_elements <- 100

base_df <- data.frame()

base_df <- colnames(c('value', 'PPC_name', 'PPC_rank'))

for(i in 1:number_elements){
        message("processing ", counts_w_replicons_hit$primary_cluster_id[i], " now, number ", rownames(PPC_by_count)[i], " of ", number_elements)
        name_PPC <- PPC_by_count$primary_cluster_id[i]
        df_temp <- mobtyper_df %>%
                filter(primary_cluster_id %in% counts_w_replicons_hit$primary_cluster_id[i]) %>%
                mutate(`Sample 1` = name) %>%
                left_join(cgMLST_dists2, by = "Sample 1")
        df_temp <- mobtyper_df %>%
                filter(primary_cluster_id %in% counts_w_replicons_hit$primary_cluster_id[i]) %>%
                mutate(`Sample 2` = name) %>%
                left_join(df_temp, by = "Sample 2")
        PPC_df <- df_temp %>% select(value, `Sample 1`, `Sample 2`, Revised_Source_Niche.x, Revised_Source_Niche.y, Consensus_phylogroup.x, Consensus_phylogroup.y, name_PPC) %>% mutate(PPC_name = name_PPC) %>% mutate(PPC_rank = i)
        base_df <- rbind(base_df, PPC_df)
        
}

Inc_types <- potential_plasmids %>% filter(primary_cluster_id %in% counts_w_replicons_hit$primary_cluster_id[1:number_elements]) %>% select(primary_cluster_id, Majority_replicon_types) %>% unique() %>% rename('PPC_name' = primary_cluster_id) %>% select(PPC_name, everything())

base_df <- Inc_types%>% right_join(base_df, by = "PPC_name")



base_df$combined_source <- apply(base_df[, c("Revised_Source_Niche.x", "Revised_Source_Niche.y")],1,function(x){
  paste(sort(x),collapse = " & ")
})

combined_source_names <- base_df %>% pull(combined_source) %>% unique() %>% sort()

combined_source_cols <- c('#e6194b', '#3cb44b', '#ffe119', '#4363d8', '#f58231', '#911eb4', '#46f0f0', '#f032e6', '#bcf60c', '#fabebe', '#008080', '#e6beff', '#9a6324', '#fffac8', '#800000', '#aaffc3', '#808000', '#ffd8b1', '#000075', '#808080', '#ffffff', '#000000')[1:length(combined_source_names)]

names(combined_source_cols) <- combined_source_names

set.seed(1)
p3 <- base_df %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y) %>% filter(!grepl("Human", combined_source)) %>% filter(PPC_name %in% counts_w_replicons_hit$primary_cluster_id[1:12]) %>% slice_sample(prop = 0.25) %>% ggplot() +
        geom_jitter(aes(x = value, y = paste(Majority_replicon_types, PPC_name), color = combined_source), alpha = 0.5) +
        theme(axis.text.x=element_text(angle=90,hjust=1)) +
        #95% of ST_new have allelic distances above this number
        #temp <- cgMLST_dists2 %>% filter(Consensus_phylogroup.x == Consensus_phylogroup.y)
        #quantile(temp$value, 0.5)
        geom_vline(xintercept = 2037, linetype = "dotted", color = "green", linewidth = 1) +
        #95% of ST_new have allelic distances below this number
        #temp <- cgMLST_dists2 %>% filter(ST_new.x == ST_new.y)
        #quantile(temp$value, 0.95)
        geom_vline(xintercept = 702, linetype = "dotted", color = "red", linewidth = 1) +
        scale_color_manual(values = combined_source_cols) 

set.seed(1)
p4 <- base_df %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y) %>% filter(grepl("Human", combined_source)) %>% filter(PPC_name %in% counts_w_replicons_hit$primary_cluster_id[1:12]) %>% slice_sample(prop = 0.25) %>% ggplot() +
        geom_jitter(aes(x = value, y = paste(Majority_replicon_types, PPC_name), color = combined_source), alpha = 0.5) +
        theme(axis.text.x=element_text(angle=90,hjust=1)) +
        #95% of ST_new have allelic distances above this number
        #temp <- cgMLST_dists2 %>% filter(Consensus_phylogroup.x == Consensus_phylogroup.y)
        #quantile(temp$value, 0.5)
        geom_vline(xintercept = 2037, linetype = "dotted", color = "green", linewidth = 1) +
        #95% of ST_new have allelic distances below this number
        #temp <- cgMLST_dists2 %>% filter(ST_new.x == ST_new.y)
        #quantile(temp$value, 0.95)
        geom_vline(xintercept = 702, linetype = "dotted", color = "red", linewidth = 1) +
        scale_color_manual(values = combined_source_cols) 



p3 + facet_wrap(combined_source ~ ., ncol = 5, nrow = 2)

p4 + facet_wrap(combined_source ~ ., ncol = 3, nrow = 2)        
     
set.seed(1)   
base_df %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y) %>% filter(grepl("F", Majority_replicon_types)) %>% slice_sample(prop = 0.25) %>% ggplot() +
        geom_jitter(aes(x = value, y = paste(Majority_replicon_types, PPC_name), color = combined_source), alpha = 0.5) +
        theme(axis.text.x=element_text(angle=90,hjust=1)) +
        #95% of ST_new have allelic distances above this number
        #temp <- cgMLST_dists2 %>% filter(Consensus_phylogroup.x == Consensus_phylogroup.y)
        #quantile(temp$value, 0.5)
        geom_vline(xintercept = 2037, linetype = "dashed", color = "green", linewidth = 1) +
        #95% of ST_new have allelic distances below this number
        #temp <- cgMLST_dists2 %>% filter(ST_new.x == ST_new.y)
        #quantile(temp$value, 0.95)
        geom_vline(xintercept = 702, linetype = "dashed", color = "red", linewidth = 1) +
        scale_color_manual(values = combined_source_cols) +
                geom_vline(xintercept = 100, linetype = "dashed", color = "black", linewidth = 1) +
        scale_color_manual(values = combined_source_cols) 

set.seed(1)
base_df %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y) %>% filter(grepl("F", Majority_replicon_types)) %>% filter(value <= 200) %>% slice_sample(prop = 0.5) %>% ggplot() +
        geom_jitter(aes(x = value, y = paste(Majority_replicon_types, PPC_name), color = combined_source), alpha = 0.5) +
        theme(axis.text.x=element_text(angle=90,hjust=1)) +
        geom_vline(xintercept = 100, linetype = "dashed", color = "black", linewidth = 1) +
        scale_color_manual(values = combined_source_cols) 


set.seed(1)
base_df %>% filter(grepl("F", Majority_replicon_types)) %>% mutate(Same_source = if_else(Revised_Source_Niche.x == Revised_Source_Niche.y, "Same Source", "Different Source")) %>% slice_sample(prop = 0.5) %>% ggplot() +
        geom_jitter(aes(x = value, y = paste(Majority_replicon_types, PPC_name), color = combined_source), alpha = 0.5) +
        theme(axis.text.x=element_text(angle=90,hjust=1, size = 8), axis.text.y=element_text(size = 8)) +
        #95% of ST_new have allelic distances above this number
        #temp <- cgMLST_dists2 %>% filter(Consensus_phylogroup.x == Consensus_phylogroup.y)
        #quantile(temp$value, 0.5)
        geom_vline(xintercept = 2037, linetype = "dashed", color = "green", linewidth = 1) +
        #95% of ST_new have allelic distances below this number
        #temp <- cgMLST_dists2 %>% filter(ST_new.x == ST_new.y)
        #quantile(temp$value, 0.95)
        geom_vline(xintercept = 702, linetype = "dashed", color = "red", linewidth = 1) +
        scale_color_manual(values = combined_source_cols) +
        geom_vline(xintercept = 100, linetype = "dashed", color = "black", linewidth = 1) +
        scale_color_manual(values = combined_source_cols) +
        guides(color=guide_legend(ncol=1))


```

## Plasmidless_distance_analyses
```{r}     

set.seed(1)
cgMLST_dists2 %>%
        filter(Consensus_phylogroup.x == Consensus_phylogroup.y & Consensus_phylogroup.x %in% c("A", "B1", "B2", "C", "D", "E", "F", "G")) %>%
        mutate(Same_source = if_else(Revised_Source_Niche.x == Revised_Source_Niche.y, "Same Source", "Different Source")) %>%
        mutate(Same_ST = if_else(ST_new.x == ST_new.y, "Same ST", "Different ST")) %>% slice_sample(prop = 0.25) %>%
        ggplot() +
        geom_violin(aes(x = value, y = Consensus_phylogroup.x), alpha = 0.5) +
        theme(axis.text.x=element_text(angle=90,hjust=1, size = 8), axis.text.y=element_text(size = 8)) +
    #95% of ST_new have allelic distances above this number
    #temp <- cgMLST_dists2 %>% filter(Consensus_phylogroup.x == Consensus_phylogroup.y)
    #quantile(temp$value, 0.5)
        geom_vline(xintercept = 2037, linetype = "dashed", color = "green", linewidth = 1) +
    #95% of ST_new have allelic distances below this number
    #temp <- cgMLST_dists2 %>% filter(ST_new.x == ST_new.y)
    #quantile(temp$value, 0.95)
        geom_vline(xintercept = 702, linetype = "dashed", color = "red", linewidth = 1) +
        geom_vline(xintercept = 100, linetype = "dashed", color = "black", linewidth = 1) + facet_wrap(Same_ST ~ ., ncol = 1)

small_df <- cgMLST_dists2 %>% filter(Consensus_phylogroup.x != Consensus_phylogroup.y)    
    
wide_plasmid_cluster_carriage <- mobtyper_df %>%
        filter(name %in% allelic_distances$name) %>%
        filter(primary_cluster_id %in% PPC_by_count$primary_cluster_id[1:20]) %>%
        select(name, primary_cluster_id) %>%
        mutate(val = 1) %>%
        pivot_wider(names_from = primary_cluster_id, values_from = val, values_fill = 0)

rm(allelic_distances)

cgMLST_dists3 <- wide_plasmid_cluster_carriage %>%
        rename("Sample 1" = name) %>% select(rev(colnames(.))) %>%
        inner_join(small_df, by = "Sample 1")

cgMLST_dists3 <- wide_plasmid_cluster_carriage %>%
        rename("Sample 2" = name) %>% inner_join(cgMLST_dists3, by = "Sample 2")

cgMLST_dists3<- cgMLST_dists3 %>% select(matches("\\.x"), `Sample 1`, value, `Sample 2`, matches("\\.y")) %>% select(!matches("\\.y"), `Sample 1`, `Sample 2`, matches("^[^A].*\\.y"), matches("^AA.*\\.y"))




df2 <- cgMLST_dists3 %>% subset(value < quantile(cgMLST_dists3$value, 0.05))

cgMLST_dists3

p1 <- ggplot() +
    geom_boxplot(data = cgMLST_dists2, aes(x = Consensus_phylogroup.x == Consensus_phylogroup.y, y = value, fill = "100%"), color = 'black') +
    #geom_violin(data = df2, aes(x = 1, y = value, fill = "95%"), color = NA) +
        scale_y_continuous(limits = c(1, 2513)) +
    #theme_classic() +
    scale_fill_grey(name = 'level')

p2 <- ggplot() +
    geom_boxplot(data = cgMLST_dists2, aes(x = ST_new.x == ST_new.y, y = value, fill = "100%"), color = 'lightblue') +
    #geom_violin(data = df2, aes(x = 1, y = value, fill = "95%"), color = NA) +
        scale_y_continuous(limits = c(1, 2513)) +
    #theme_classic() +
    scale_fill_grey(name = 'level')
```

```{r}

ska_dists <- vroom("analysis/TEA/ska_manual/")

cgMLST_less_100 <- cgMLST_dists2 %>% filter(value <= 100)

cgMLST_less_100_w_plas <- mobtyper_df %>% rename(`Sample 1` = name) %>% select(`Sample 1`, primary_cluster_id, secondary_cluster_id) %>% inner_join(cgMLST_less_100, by = "Sample 1")

cgMLST_less_100_w_plas <- mobtyper_df %>% rename(`Sample 2` = name) %>% select(`Sample 2`, primary_cluster_id, secondary_cluster_id) %>% inner_join(cgMLST_less_100_w_plas, by = "Sample 2")

SKA_ST963 <- read_delim("analysis/TEA/ska_manual/ST963_dists/ST963_dist_100_SNP.distances.tsv", delim = "\t")

SKA_ST963 %>% left_join(cgMLST_less_100_w_plas, by = c("Sample 1", "Sample 2"))
```



```{r phylogroup_tree, echo=FALSE}
#Overwrite with different colours
source_cols <- c("Companion Animal" = "#59398d",
         "Environmental" = "#709b46",
         "Food" = "#c09f3d",
         "Human" = "#48c595",
         "Livestock" =  "#c26bbc",
         "Wild Animal" = "#b9553d")

#Create fake colour for header
phylogroup_header <- c(NA)
names(phylogroup_header) <- 'Phylogroups'

#Define our colours for our phylogroups
clermont_cols <- c('#ffe119', '#4363d8', '#f58231', '#dcbeff', '#800000', '#000075', '#9A6324', 'red', '#3cb44b',  'white', 'grey60', 'grey30', 'black')

#Assign our color names for phylogroups
names(clermont_cols) <- c("A", "B1",  "B2", "C", "D", "E", "F", "G", "E or cladeI", "cladeI", "cladeIII", "cladeIV", "Unknown")

#Combine our lists of colours
var_cols <- c(source_cols, clermont_cols)

#Create the tree
circ_tree <- ggtree(tree,
        layout = 'circular',
        open.angle = 0) %<+%
        genometa +
        #geom_tiplab(size = 0.25,
        #        align = TRUE,
        #        linetype = NULL,
        #        aes(color = as.factor(Revised_Source_Niche))
        #        ) +
        #geom_tiplab(size = 0.25,
        #        align = TRUE,
        #        linetype = NULL,
        #        offset = 200,
        #        aes(label = mlst..ST)
        #        ) +
        geom_tippoint(size = 0.5,
                align = TRUE,
                linetype = NULL,
                aes(color = Consensus_phylogroup)
                ) + 
        guides(color = guide_legend(override.aes = list(size = 5)), shape = guide_legend(override.aes = list(size = 5)))

tree_df <- mobtyper_df %>% left_join(sample_names, by = "name") %>% select(name, primary_cluster_id) %>% mutate(val = 1) %>% pivot_wider(names_from = "primary_cluster_id", values_from = "val", values_fill = 0) %>% select(name, all_of(counts_w_replicons_hit$primary_cluster_id[1:20])) %>% column_to_rownames('name') %>% mutate(across(everything(), ~ as.character(.x)))

p3 <- gheatmap(
        p = circ_tree,
        data = tree_df,
        colnames_offset_y = -0.1,
        font.size = 1.5,
        hjust = 0,
        colnames = TRUE,
        offset = 1,
        width = 0.3,
        color = NULL
        ) +
theme(legend.position = "right") +#,
        #legend.box = "vertical",
        #legend.key.size = unit(1, "mm"),
        #legend.title = element_blank(),
        #legend.text = element_text(size = 8)
        #) +
scale_fill_manual(
        values = c(source_cols, clermont_cols, "1" = "black", "0" = "white"),
        na.value = 'grey'
        )  +
scale_colour_manual(
        values = clermont_cols,
        guide = "none"
        )# +
        #ylim(NA, 100)

ggsave("bigtreewplas.pdf", plot = p3, width = 11.7, height = 8.3)

tree_df %>% rownames_to_column("strain") %>% write_csv("delims/plasmids_tree_df.txt")

```


# Explore contig-wise mobtyper data
```{r}
heatmap_df <- counts_w_replicons_hit %>% filter(primary_cluster_id %in% decat_known_PPCs_known_rep_wide$primary_cluster_id)  %>% select(primary_cluster_id, 10:ncol(.))%>% column_to_rownames("primary_cluster_id")

colnames(heatmap_df) <- paste0(colnames(heatmap_df), " n = ",colSums(heatmap_df),"/", nrow(heatmap_df))

heatmap_df %>% select(matches("Inc")) %>% filter(rowSums(.) != 0) %>% pheatmap::pheatmap(fontsize_row = 2)

heatmap_df %>% select(matches("Col")) %>% filter(rowSums(.) != 0) %>% pheatmap::pheatmap(fontsize_row = 2)

pheatmap_df <- clusters %>% filter(primary_cluster_id %in% decat_known_PPCs_known_rep_wide$primary_cluster_id) %>% filter(`IncF RST` != "F-:A-:B-") %>% select(primary_cluster_id, matches("ABRICATE..card")) %>% group_by(primary_cluster_id) %>% mutate(across(everything(), ~ mean(.))) %>% unique() %>% column_to_rownames('primary_cluster_id') %>% mutate(across(everything(), ~ replace_na(., 0))) %>% filter(rowSums(.) != 0) %>% select(where(~ is.numeric(.x) && sum(.x) !=0 ))

p1 <- pheatmap_df %>% pheatmap::pheatmap(fontsize_row = 2, fontsize_col = 1)


#Read in abricate data for tea dataset
abricateR::abricateR(abricate_in = "/Users/131785/Dropbox/PostDoc/Austrakka/AusTrakka_R/analysis/TEA/genotype.txt", output = "TEA_abricate", writecsv = FALSE)


mobtyper_contig_wise <- read_delim("analysis/TEA/mobtyper_results.txt", 
    delim = "\t", escape_double = FALSE, 
    trim_ws = TRUE)


#Read in our mobtyper data
mobtyper_contig_wise <- mobtyper_contig_wise %>% rename('name' = sample_id) %>% filter(name %in% tree$tip.label)

small_mobtyper_hits <- mobtyper_contig_wise %>% select(name, contig_id, primary_cluster_id) %>% mutate(contig_id = str_replace(contig_id, " .*", ""))

co_occurence_df <- TEA_abricate_co_occurence_N90L90 %>% rename('contig_id' = SEQUENCE) %>% left_join(small_mobtyper_hits, by = c("name", "contig_id"))

co_occurence_df_wide <- co_occurence_df %>%
        group_by(name, primary_cluster_id) %>%
        summarise(same_scaff = paste0(same_scaff, collapse = " ")) %>%
        #filter(primary_cluster_id %in% decat_known_PPCs_known_rep_wide$primary_cluster_id, primary_cluster_id != is.na(primary_cluster_id)) %>%
        mutate(same_scaff2 = strsplit(same_scaff, " ")) %>%
     unnest(same_scaff2) %>% 
     arrange(same_scaff2) %>%  
     pivot_wider(names_from = same_scaff2,
            names_repair = "universal", values_from = same_scaff2, 
        values_fill = 0, values_fn = length)




pheatmap_df2 <- co_occurence_df_wide %>% group_by(primary_cluster_id) %>% add_count() %>% ungroup() %>% select(primary_cluster_id, matches("card")) %>% group_by(primary_cluster_id) %>% mutate(across(everything(), ~ mean(.))) %>% unique() %>% filter(primary_cluster_id %in% rownames(pheatmap_df)) %>% column_to_rownames('primary_cluster_id') %>% mutate(across(everything(), ~ replace_na(., 0))) %>% select(where(~ is.numeric(.x) && sum(.x) !=0 )) 



pheatmap_df_match_other <- pheatmap_df[rownames(pheatmap_df) %in% rownames(pheatmap_df2),]

rownames(pheatmap_df2) <- paste(rownames(pheatmap_df2), "Number 2")


pheatmap_df_match_other <- pheatmap_df_match_other %>% rownames_to_column('name') 

colnames(pheatmap_df_match_other) <- gsub("ABRICATE\\.\\.", "", colnames(pheatmap_df_match_other))

pheatmap_df2 <- pheatmap_df2 %>% rownames_to_column('name')

full_join(pheatmap_df_match_other, pheatmap_df2) %>% column_to_rownames('name') %>% mutate(across(everything(), ~ replace_na(., 0))) %>% pheatmap::pheatmap(fontsize_row = 2)

```




```{r}


#Join our data on our IncF RST typeable plasmid clusters with our mobtyper data for our genomes
Samples_w_matched_IncF_clusters <- inner_join(mobtyper_df, IncF_RST_pos_clusters, by = "mash_nearest_neighbor")

#Create a list of samples with typeable F plasmids which share a plasmid clustter and an IncF RST with  typeable F plasmids from the plasmid db
IncF_typeable_samples_w_matched_IncF_clusters <- Samples_matched_IncF_clusters %>% filter(`IncF RST.x` != "F-:A-:B-") %>% mutate(same_IncF_RST = `IncF RST.x` == `IncF RST.y`) %>% select(name, `IncF RST.x`, `IncF RST.y`, same_IncF_RST, size.x, size.y, everything()) %>% filter(same_IncF_RST == TRUE)

#Create a list of samples with NO typeable F plasmids that share a plasmid cluster with F plasmids from the plasmid db
no_matching_fs <- Samples_matched_IncF_clusters %>% 
        filter(name %nin% IncF_typeable_samples_w_matched_IncF_clusters$name)

#Create a list of samples with typeable F plasmids which share a plasmid cluster but not an IncF RST with  typeable F plasmids from the plasmid db
Samples_matched_IncF_clusters %>%
        mutate(same_IncF_RST = `IncF RST.x` == `IncF RST.y`) %>%
        filter(same_IncF_RST == FALSE) %>%
        filter(`IncF RST.x`!= "F-:A-:B-") %>%
        mutate(Fbit1 = grepl("IncF", `rep_type(s).x`)) %>%
        mutate(Fbit2 = grepl("F", `relaxase_type(s).x`)) %>%
        mutate(Fbit3 = grepl("F", mpf_type.x)) %>%
        mutate(Fbit4 = grepl("F", `orit_type(s).x`)) %>%
        mutate(sum_of_f_bits = rowSums(.[,c('Fbit1', 'Fbit2', 'Fbit3', 'Fbit4')])) %>%
        group_by(name) %>%
        filter(max(sum_of_f_bits) == sum_of_f_bits, max(size.x) == size.x) %>%
        mutate(size_diff = size.x/size.y) %>% select(size_diff, sum_of_f_bits, everything())

has_f <- has_matching_f %>% inner_join(sources, by = "name") %>% pull(Revised_Source_Niche) %>% table()

unclean_f <- no_matching_fs %>% inner_join(sources, by = "name") %>% pull(Revised_Source_Niche) %>% table()

no_f <- left_join(mobtyper_df, IncF_RST_pos_clusters, by = "mash_nearest_neighbor") %>% filter(`IncF RST.x` == "F-:A-:B-") %>% inner_join(sources, by = "name") %>% pull(Revised_Source_Niche) %>% table()

total_n <- sources %>% pull(Revised_Source_Niche) %>% table()

cbind(has_f, unclean_f, no_f, total_n)

Samples_matched_IncF_clusters %>%
        arrange(size.x) %>%
        group_by(name) %>%
        mutate(PPC_F = case_when(
               same_IncF_RST == TRUE ~  primary_cluster_id.x,
               same_IncF_RST == FALSE ~ largest_cluster_w_match, 
        ))
```