---
title: "AusTrakka Basic Accessory Exploration"
author: "Max Cummins"
date: "`r Sys.Date()`"
output:
    html_document:
        theme: spacelab
        toc: true
        toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#Read in our packages
library(tidyverse)
library(vroom)
library(ggplot2)
library(xml2)
library(ggrepel)
library(caret)
library(pheatmap)
library(ComplexHeatmap)
library(lme4)
library(parallel)
library(circlize)

#Define our not in function
`%nin%` <- Negate(`%in%`)
```

```{r}
#Overwrite with different colours
source_cols <- c("Companion Animal" = "#59398d",
         "Environmental" = "#709b46",
         "Food" = "#c09f3d",
         "Human" = "#48c595",
         "Livestock" =  "#c26bbc",
         "Wild Animal" = "#b9553d")

#Define our colours for our phylogroups
clermont_cols <- c('#ffe119', '#4363d8', '#f58231', '#dcbeff', '#800000', '#000075', '#9A6324', 'red', '#3cb44b',  'white', '#a9a9a9', 'black')

#Assign our color names for phylogroups
names(clermont_cols) <- c("A", "B1",  "B2", "C", "D", "E", "F", "G", "E or cladeI", "cladeI", "cladeIII", "cladeIV")
```


# MobSuite Processing

Below we will explore our plasmid data as generated by MobSuite.

## Read in our tree and genomic and metadata dataset

```{r Merge_analysis_outputs, include=FALSE}
genometa <- vroom("delims/genometa_n5631.txt", show_col_types = FALSE)

genometa <- genometa %>% mutate(Revised_Source_Niche = factor(Revised_Source_Niche,
                                                  levels = c("Companion Animal",
                                                             "Environmental",
                                                             "Food",
                                                             "Human",
                                                             "Livestock" ,
                                                             "Wild Animal")))

genometa <- genometa %>% mutate(Consensus_phylogroup = factor(Consensus_phylogroup,
                                                              levels = c("A",
                                                                         "B1",
                                                                         "B2",
                                                                         "C",
                                                                         "D",
                                                                         "E",
                                                                         "F",
                                                                         "G",
                                                                         "E or cladeI",
                                                                         "cladeI",
                                                                         "cladeIII",
                                                                         "cladeIV",
                                                                         "Unknown")))

#Read in our tree
tree <- ggtree::read.tree("analysis/TEA/cgmlst_coreugate/pad.nwk")

#Remove quotes from sample names
tree$tip.label <- gsub("'", "", tree$tip.label)

#Create a list of samples in our cohort
sample_names <- tree$tip.label %>% as.data.frame() %>% rename('name' = '.')

#Read in our abritamr data
long_abritamr <- vroom("delims/long_format_abritamr_5631.txt", show_col_types = FALSE)
```

```{r}
small_meta <- genometa %>% select(name, Consensus_phylogroup, ST_new, Revised_Source_Niche)

long_abritamr_w_meta <- full_join(small_meta, long_abritamr, by = "name")

long_abritamr_w_meta %>% select(name, Strain_count_of_AMR_genes, Consensus_phylogroup, Revised_Source_Niche, Strain_CIA_status) %>% unique() %>% ggplot() +
    geom_jitter(aes(x = Strain_count_of_AMR_genes, y = Consensus_phylogroup, color = Revised_Source_Niche, shape = as.character(Strain_CIA_status)), size = 0.75) +
    geom_boxplot(aes(x = Strain_count_of_AMR_genes, y = Consensus_phylogroup), alpha = 0.25)+
    scale_y_discrete(limits=rev) +
        scale_x_continuous(breaks = seq(min(long_abritamr_w_meta$Strain_count_of_AMR_genes), max(long_abritamr_w_meta$Strain_count_of_AMR_genes), by = 1)) +
    scale_colour_manual(values = source_cols)

#Plot Source vs classwise res
long_abritamr_w_meta %>% select(name, Revised_Source_Niche, resistance, resistance_genes_present) %>% pivot_wider(names_from = "resistance", values_from = "resistance_genes_present", values_fill = 0) %>% group_by(Revised_Source_Niche) %>% summarise(across(2:35, mean)) %>% column_to_rownames("Revised_Source_Niche") %>% select(-`NA`) %>% pheatmap(display_numbers = T, color = colorRamp2(c(0, 1), c("white", "red")), legend = T, cellheight = 10, cluster_cols = F, cellwidth = 20)

#Plot Source vs CIA/MDR
long_abritamr_w_meta %>% select(name, Revised_Source_Niche, Strain_CIA_status, Strain_MDR_status) %>% unique() %>% group_by(Revised_Source_Niche) %>% summarise(across(2:3, mean))  %>% mutate(Revised_Source_Niche = factor(Revised_Source_Niche,
                                                  levels = c("Environmental",
                                                             "Human",
                                                             "Livestock" ,
                                                             "Wild Animal",
                                                             "Companion Animal",
                                                             "Food"))) %>% arrange(Revised_Source_Niche) %>% column_to_rownames("Revised_Source_Niche") %>% pheatmap(display_numbers = T, color = colorRamp2(c(0, 1), c("white", "red")), legend = F, cellheight = 10, cluster_cols = F, cluster_rows = F, cellwidth = 20)

#Plot Phylogroup vs classwise res
long_abritamr_w_meta %>% select(name, Consensus_phylogroup, resistance, resistance_genes_present) %>% pivot_wider(names_from = "resistance", values_from = "resistance_genes_present", values_fill = 0) %>% group_by(Consensus_phylogroup) %>% summarise(across(2:35, mean)) %>% column_to_rownames("Consensus_phylogroup") %>% select(-`NA`) %>% pheatmap(display_numbers = T, color = colorRamp2(c(0, 1), c("white", "red")), legend = F, cellheight = 10, cluster_cols = F, cluster_rows = F, cellwidth = 20)

#Plot Phylogroup vs CIA/MDR
long_abritamr_w_meta %>% select(name, Consensus_phylogroup, Strain_CIA_status, Strain_MDR_status)  %>% unique() %>% group_by(Consensus_phylogroup) %>% summarise(across(2:3, mean)) %>% column_to_rownames("Consensus_phylogroup") %>% pheatmap(display_numbers = T, color = colorRamp2(c(0, 1), c("white", "red")), legend = F, cellheight = 10, cluster_cols = F, cluster_rows = F, cellwidth = 20)

#Plot Total classwise res
long_abritamr_w_meta %>% select(name, resistance, resistance_genes_present) %>% pivot_wider(names_from = "resistance", values_from = "resistance_genes_present", values_fill = 0) %>% summarise(across(2:35, mean)) %>% select(-`NA`) %>% mutate(rownem = "Total") %>% column_to_rownames("rownem") %>% pheatmap(display_numbers = T, color = colorRamp2(c(0, 1), c("white", "red")), legend = F, cellheight = 10, cluster_cols = F, cluster_rows = F, cellwidth = 20)

#Plot Total vs CIA/MDR
long_abritamr_w_meta %>% select(name, Strain_CIA_status, Strain_MDR_status) %>% unique() %>% summarise(across(2:3, mean)) %>% mutate(rownem = "Total") %>% column_to_rownames("rownem") %>% pheatmap(display_numbers = T, color = colorRamp2(c(0, 1), c("white", "red")), legend = F, cellheight = 10, cluster_cols = F, cluster_rows = F, cellwidth = 20)


```

```{r}
MDR_CIA_res_phylogroup <- long_abritamr_w_meta %>% select(name, Consensus_phylogroup, Strain_CIA_status, Strain_MDR_status) %>% unique() %>% filter(Consensus_phylogroup %in% c('A', 'B1', 'B2', 'C', 'D', 'E', 'F', 'G')) 

MDR_CIA_res_phylogroup_aov <- aov(Strain_CIA_status ~ Consensus_phylogroup, data = MDR_CIA_res_phylogroup)

summary(MDR_CIA_res_phylogroup_aov)

tukey_MDR_CIA_res_phylogroup_aov <- TukeyHSD(MDR_CIA_res_phylogroup_aov)

MDR_res_phylogroup_aov <- aov(Strain_MDR_status ~ Consensus_phylogroup, data = MDR_CIA_res_phylogroup)

summary(MDR_res_phylogroup_aov)

tukey_MDR_res_phylogroup_aov <- TukeyHSD(MDR_res_phylogroup_aov)








MDR_CIA_res_source <- long_abritamr_w_meta %>% select(name, Revised_Source_Niche, Strain_CIA_status, Strain_MDR_status)  %>% unique() 


MDR_CIA_res_source_aov <- aov(Strain_CIA_status ~ Revised_Source_Niche, data = MDR_CIA_res_source)

summary(MDR_CIA_res_source_aov)

posthoc <- TukeyHSD(MDR_CIA_res_source_aov)

MDR_CIA_res_source <- long_abritamr_w_meta %>% select(name, Revised_Source_Niche, Strain_CIA_status, Strain_MDR_status)  %>% unique() 


MDR_res_source_aov <- aov(Strain_MDR_status ~ Revised_Source_Niche, data = MDR_CIA_res_source)

summary(MDR_res_source_aov)

posthoc <- TukeyHSD(MDR_res_source_aov)


```

```{r}

MDR_CIA_res_phylogroup <- long_abritamr_w_meta %>% select(name, Consensus_phylogroup, Revised_Source_Niche, Strain_CIA_status, Strain_MDR_status, Strain_count_of_AMR_genes, Strain_count_of_AMR_gene_classes) %>% unique() %>% filter(Consensus_phylogroup %in% c('A', 'B1', 'B2', 'C', 'D', 'E', 'F', 'G')) 

library(ggpubr)

comparisons <- combn(as.character(sort(unique(MDR_CIA_res_phylogroup$Consensus_phylogroup))), 2, simplify = F)

p <- ggboxplot(MDR_CIA_res_phylogroup, x = "Consensus_phylogroup", y = "Strain_count_of_AMR_gene_classes", fill = "Consensus_phylogroup", palette = clermont_cols)
#  Add p-value
#p + stat_compare_means(comparisons = comparisons, label = "p.signif")
p + stat_compare_means(method = "anova", label.y = 11) +
        stat_compare_means(label = "p.signif", method = "t.test", ref.group = ".all.") + 
        geom_hline(yintercept = mean(MDR_CIA_res_phylogroup$Strain_count_of_AMR_gene_classes), linetype = 2)



comparisons <- combn(as.character(sort(unique(MDR_CIA_res_phylogroup$Revised_Source_Niche))), 2, simplify = F)

p <- ggboxplot(MDR_CIA_res_phylogroup, x = "Revised_Source_Niche", y = "Strain_count_of_AMR_gene_classes", fill = "Revised_Source_Niche", palette = source_cols)
#  Add p-value
#p + stat_compare_means(comparisons = comparisons, label = "p.signif")
p + stat_compare_means(method = "anova", label.y = 11) +
        stat_compare_means(label = "p.signif", method = "t.test", ref.group = ".all.") + 
        geom_hline(yintercept = mean(MDR_CIA_res_phylogroup$Strain_count_of_AMR_gene_classes), linetype = 2)
```


```{r}
MDR_CIA_res_phylogroup <- long_abritamr_w_meta %>% select(name, Consensus_phylogroup, Strain_CIA_status, Strain_MDR_status) %>% unique() %>% filter(Consensus_phylogroup %in% c('A', 'B1', 'B2', 'C', 'D', 'E', 'F', 'G')) 

MDR_CIA_res_phylogroup_aov <- aov(Strain_CIA_status ~ Consensus_phylogroup, data = MDR_CIA_res_phylogroup)

summary(MDR_CIA_res_phylogroup_aov)

tukey_MDR_CIA_res_phylogroup_aov <- TukeyHSD(MDR_CIA_res_phylogroup_aov)

MDR_res_phylogroup_aov <- aov(Strain_MDR_status ~ Consensus_phylogroup, data = MDR_CIA_res_phylogroup)

summary(MDR_res_phylogroup_aov)

tukey_MDR_res_phylogroup_aov <- TukeyHSD(MDR_res_phylogroup_aov)








MDR_CIA_res_source <- long_abritamr_w_meta %>% select(name, Revised_Source_Niche, Strain_CIA_status, Strain_MDR_status)  %>% unique() 


MDR_CIA_res_source_aov <- aov(Strain_CIA_status ~ Revised_Source_Niche, data = MDR_CIA_res_source)

summary(MDR_CIA_res_source_aov)

posthoc <- TukeyHSD(MDR_CIA_res_source_aov)

MDR_CIA_res_source <- long_abritamr_w_meta %>% select(name, Revised_Source_Niche, Strain_CIA_status, Strain_MDR_status)  %>% unique() 


MDR_res_source_aov <- aov(Strain_MDR_status ~ Revised_Source_Niche, data = MDR_CIA_res_source)

summary(MDR_res_source_aov)

posthoc <- TukeyHSD(MDR_res_source_aov)


```

```{r}
#long_abritamr_w_meta %>% select(name, resistance, class_includes_CIA_res, resistance_genes_present) %>% unique() %>% filter(class_includes_CIA_res == 1)  %>% group_by(resistance) %>% full_join(small_meta) %>% pivot_wider(values_from = "resistance_genes_present", names_from = "resistance", values_fill = 0) %>% full_join(sample_names) %>% mutate(across(3:ncol(.), ~ replace_na(., 0))) %>% select(-`NA`) %>% group_by(Revised_Source_Niche) %>% summarise(across(6:16), mean)
```


```{r}

vf_count_phylogroup <- genometa %>% select(name, Consensus_phylogroup, Revised_Source_Niche, Strain_VF_count) %>% filter(Consensus_phylogroup %in% c('A', 'B1', 'B2', 'C', 'D', 'E', 'F', 'G'))  %>% mutate(Strain_VF_count = if_else(is.na(Strain_VF_count), 0, Strain_VF_count))


p <- ggboxplot(vf_count_phylogroup, x = "Consensus_phylogroup", y = "Strain_VF_count", fill = "Consensus_phylogroup", palette = clermont_cols)
#  Add p-value
#p + stat_compare_means(comparisons = comparisons, label = "p.signif")
p + stat_compare_means(method = "anova", label.y = 170) +
        stat_compare_means(label = "p.signif", method = "t.test", ref.group = ".all.") + 
        geom_hline(yintercept = mean(vf_count_phylogroup$Strain_VF_count), linetype = 2)




comparisons <- combn(as.character(sort(unique(genometa$Revised_Source_Niche))), 2, simplify = F)

p <- ggboxplot(vf_count_phylogroup, x = "Revised_Source_Niche", y = "Strain_VF_count", fill = "Revised_Source_Niche", palette = source_cols)
#  Add p-value
#p + stat_compare_means(comparisons = comparisons, label = "p.signif")
p + stat_compare_means(method = "anova", label.y = 170) +
        stat_compare_means(label = "p.signif", method = "t.test", ref.group = ".all.") + 
        geom_hline(yintercept = mean(vf_count_phylogroup$Strain_VF_count), linetype = 2)
```
```{r}
vf_df <- genometa %>% column_to_rownames("name") %>% select(Revised_Source_Niche, Consensus_phylogroup, matches("EC_custom"), matches("vfdb"))

#Aggregate duplicate columns
vf_df <- vf_df %>%
        mutate("new_afaE" = rowSums(across(matches("EC_custom_afaE")))) %>% select(-matches("EC_custom_afaE")) %>%
        mutate("new_irp2/fyuA" = rowSums(across(matches("EC_custom_irp2_|vfdb_irp2|EC_custom_fyuA")))) %>% select(-matches("EC_custom_irp2_|vfdb_irp2|EC_custom_fyuA")) %>%
        mutate("new_iutA/iuc*" = rowSums(across(matches("EC_custom_iutA_|vfdb_iutA|vfdb_iuc")))) %>% select(-matches("EC_custom_iutA_|vfdb_iutA|vfdb_iuc")) %>%
        mutate("new_cjr/senB*" = rowSums(across(matches("EC_custom_cjr|EC_custom_senB_|vfdb_senB")))) %>% select(-matches("EC_custom_cjr|EC_custom_senB_|vfdb_senB")) %>%
        mutate("new_lhr" = rowSums(across(matches("EC_custom_lhr_")))) %>% select(-matches("EC_custom_lhr_")) %>%
        mutate("new_sit*" = rowSums(across(matches("EC_custom_sitA_")))) %>% select(-matches("EC_custom_sitA_")) %>%
        mutate("new_tia" = rowSums(across(matches("EC_custom_tia_")))) %>% select(-matches("EC_custom_tia_")) %>%
        mutate("new_usp" = rowSums(across(matches("EC_custom_usp_")))) %>% select(-matches("EC_custom_usp_")) %>%
        mutate("new_pap*" = rowSums(across(matches("EC_custom_papG|vfdb_pap")))) %>% select(-matches("EC_custom_pap|vfdb_pap")) %>%
        mutate("new_kpsMT*" = rowSums(across(matches("EC_custom_kps|vfdb_kps)")))) %>% select(-matches("EC_custom_kps|vfdb_kps")) %>%
        mutate("new_ybt*" = rowSums(across(matches("vfdb_ybt")))) %>% select(-matches("vfdb_ybt")) %>%
        mutate("new_yag/ecp*" = rowSums(across(matches("vfdb_yag")))) %>% select(-matches("vfdb_yag")) %>%
        mutate("new_bfp*" = rowSums(across(matches("vfdb_bfp")))) %>% select(-matches("vfdb_bfp")) %>%
        mutate("new_neu*" = rowSums(across(matches("EC_custom_neu|vfdb_neu")))) %>% select(-matches("EC_custom_neu|vfdb_neu")) %>%
        mutate("new_aaf*" = rowSums(across(matches("vfdb_aaf")))) %>% select(-matches("vfdb_aaf")) %>%
        mutate("new_aat*" = rowSums(across(matches("vfdb_aat")))) %>% select(-matches("vfdb_aat")) %>%
        mutate("new_ces*" = rowSums(across(matches("vfdb_ces")))) %>% select(-matches("vfdb_ces")) %>%
        mutate("new_cfa*" = rowSums(across(matches("vfdb_cfa")))) %>% select(-matches("vfdb_cfa")) %>%
        mutate("new_sfa*" = rowSums(across(matches("vfdb_sfa")))) %>% select(-matches("vfdb_sfa")) %>%
        mutate("new_cgs*" = rowSums(across(matches("vfdb_cgs")))) %>% select(-matches("vfdb_cgs")) %>%
        mutate("new_clb*" = rowSums(across(matches("vfdb_clb")))) %>% select(-matches("vfdb_clb")) %>%
        mutate("new_dra*" = rowSums(across(matches("vfdb_dra")))) %>% select(-matches("vfdb_dra")) %>%
        mutate("new_chu*" = rowSums(across(matches("vfdb_chu")))) %>% select(-matches("vfdb_chu")) %>%
        mutate("new_eltA/B" = rowSums(across(matches("vfdb_elt")))) %>% select(-matches("vfdb_elt")) %>%
        mutate("new_ent*" = rowSums(across(matches("vfdb_ent")))) %>% select(-matches("vfdb_ent")) %>%
        mutate("new_esc*" = rowSums(across(matches("vfdb_esc")))) %>% select(-matches("vfdb_esc")) %>%
        mutate("new_mrk*" = rowSums(across(matches("vfdb_mrk")))) %>% select(-matches("vfdb_mrk")) %>%
        mutate("new_mxi*" = rowSums(across(matches("vfdb_mxi")))) %>% select(-matches("vfdb_mxi")) %>%
        mutate("new_nle*" = rowSums(across(matches("vfdb_nle")))) %>% select(-matches("vfdb_nle")) %>%
        mutate("new_esp*" = rowSums(across(matches("vfdb_esp")))) %>% select(-matches("vfdb_esp")) %>%
        mutate("new_tss*" = rowSums(across(matches("vfdb_tss")))) %>% select(-matches("vfdb_tss")) %>%
        mutate("new_fep*" = rowSums(across(matches("vfdb_fep")))) %>% select(-matches("vfdb_fep")) %>%
        mutate("new_spa*" = rowSums(across(matches("vfdb_spa")))) %>% select(-matches("vfdb_spa")) %>%
        mutate("new_shu*" = rowSums(across(matches("vfdb_shu")))) %>% select(-matches("vfdb_shu")) %>%
        mutate("new_sep*" = rowSums(across(matches("vfdb_sep")))) %>% select(-matches("vfdb_sep")) %>%
        mutate("new_osp*" = rowSums(across(matches("vfdb_osp")))) %>% select(-matches("vfdb_osp")) %>%
        mutate("new_iro*" = rowSums(across(matches("vfdb_iro")))) %>% select(-matches("vfdb_iro")) %>%
        mutate("new_ipg*" = rowSums(across(matches("vfdb_ipg")))) %>% select(-matches("vfdb_ipg")) %>%
        mutate("new_gsp*" = rowSums(across(matches("vfdb_gsp")))) %>% select(-matches("vfdb_gsp")) %>%
        mutate("new_foc*" = rowSums(across(matches("vfdb_foc")))) %>% select(-matches("vfdb_foc")) %>%
        mutate("new_eit*" = rowSums(across(matches("EC_custom_eit")))) %>% select(-matches("EC_custom_eit")) %>%
        mutate("new_cva*" = rowSums(across(matches("EC_custom_cva")))) %>% select(-matches("EC_custom_cva")) %>%
        mutate("new_sub*" = rowSums(across(matches("EC_custom_sub")))) %>% select(-matches("EC_custom_sub")) %>%
        mutate("new_ibeA" = rowSums(across(matches("EC_custom_ibeA")))) %>% select(-matches("EC_custom_ibeA"))

vf_df <- vf_df %>%
        select(-matches("EC_custom_VGI")) %>%
        select(-matches("EC_custom_intI")) %>%
        select(-matches("EC_custom_terA")) %>%
        select(-matches("fim[^H]")) %>%
        select(-matches("EC_custom_fimH"))
        
#gsub(".*EC_custom_|.*vfdb_", "", colnames(vf_df)) %>% sort()

vf_df <- vf_df %>% mutate(across(where(is.numeric), ~ if_else(. >= 1, 1, 0)))

vf_categories <- vf_df %>% select(-matches("ABRICATE"), -Revised_Source_Niche, -Consensus_phylogroup) %>%
        colnames() %>% sort() %>% as.data.frame() %>% rename('Gene' = '.') 

vf_categories <- vf_categories <- vf_categories %>% mutate('Function' = case_when(Gene == 'new_aaf*' ~ 'Adhesion',
                                                Gene == 'new_aat*' ~ 'Autotransporter',
                                                Gene == 'new_afaE' ~ 'Adhesion',
                                                Gene == 'new_bfp*' ~ 'Adhesion',
                                                Gene == 'new_ces*' ~ 'Chaperone',
                                                Gene == 'new_cfa*' ~ 'Adhesion',
                                                Gene == 'new_cgs*' ~ 'Adhesion',
                                                Gene == 'new_chu*' ~ 'Iron acquisition',
                                                Gene == 'new_cjr/senB*' ~ 'Bacteriocin-associated',
                                                Gene == 'new_clb*' ~ 'Toxin',
                                                Gene == 'new_cva*' ~ 'Bacteriocin-associated',
                                                Gene == 'new_dra*' ~ 'Adhesion',
                                                Gene == 'new_eit*' ~ 'Iron acquisition',
                                                Gene == 'new_eltA/B' ~ 'Toxin',
                                                Gene == 'new_ent*' ~ 'Iron acquisition',
                                                Gene == 'new_esc*' ~ 'Secretion system associated',
                                                Gene == 'new_esp*' ~ 'Secretion system associated',
                                                Gene == 'new_fep*' ~ 'Iron acquisition',
                                                Gene == 'new_foc*' ~ 'Iron acquisition',
                                                Gene == 'new_foc*' ~ 'Adhesion',
                                                Gene == 'new_gsp*' ~ 'Secretion system associated',
                                                Gene == 'new_ibeA' ~ 'Invasion',
                                                Gene == 'new_ipg*' ~ 'Secretion system associated',
                                                Gene == 'new_iro*' ~ 'Iron acquisition',
                                                Gene == 'new_irp2/fyuA' ~ 'Iron acquisition',
                                                Gene == 'new_iutA/iuc*' ~ 'Iron acquisition',
                                                Gene == 'new_kpsMT*' ~ 'Protectin',
                                                Gene == 'new_lhr' ~ 'Heat/Chlorine resistance',
                                                Gene == 'new_mrk*' ~ 'Adhesion',
                                                Gene == 'new_mxi*' ~ 'Secretion system associated',
                                                Gene == 'new_neu*' ~ 'Protectin',
                                                Gene == 'new_nle*' ~ 'Toxin',
                                                Gene == 'new_osp*' ~ 'Other',
                                                Gene == 'new_pap*' ~ 'Adhesion',
                                                Gene == 'new_sep*' ~ 'Adhesion',
                                                Gene == 'new_sfa*' ~ 'Adhesion',
                                                Gene == 'new_shu*' ~ 'Iron acquisition',
                                                Gene == 'new_sit*' ~ 'Iron acquisition',
                                                Gene == 'new_spa*' ~ 'Secretion system associated',
                                                Gene == 'new_sub*' ~ 'Toxin',
                                                Gene == 'new_tia' ~ 'Invasion',
                                                Gene == 'new_tss*' ~ 'Secretion system associated',
                                                Gene == 'new_usp' ~ 'Bacteriocin-associated',
                                                Gene == 'new_yag/ecp*' ~ 'Adhesion',
                                                Gene == 'new_ybt*' ~ 'Other'
                                                ))

vf_categories <- vf_categories %>% mutate('Name' = case_when(Gene == 'new_aaf*' ~ 'Aggregative Adhesion Fimbria',
                                            Gene == 'new_aat*' ~ 'APEC Autotransporter',
                                            Gene == 'new_afaE' ~ 'Afimbrial adhesin',
                                            Gene == 'new_bfp*' ~ 'Bundle-forming pili',
                                            Gene == 'new_ces*' ~ '',
                                            Gene == 'new_cfa*' ~ 'Colonization factor antigen',
                                            Gene == 'new_cgs*' ~ 'Curli fimbriae',
                                            Gene == 'new_chu*' ~ '',
                                            Gene == 'new_cjr/senB*' ~ 'Colicin Js receptor',
                                            Gene == 'new_clb*' ~ 'Colibactin',
                                            Gene == 'new_cva*' ~ 'Colicin V',
                                            Gene == 'new_dra*' ~ '',
                                            Gene == 'new_eit' ~ 'Iron acquisition',
                                            Gene == 'new_eltA/B' ~ 'heat-labile enterotoxin A/B',
                                            Gene == 'new_ent*' ~ 'Enterobactin',
                                            Gene == 'new_esc*' ~ '',
                                            Gene == 'new_esp*' ~ '',
                                            Gene == 'new_fep*' ~ 'Ferric enterobactin transport protein',
                                            Gene == 'new_foc*' ~ 'F1C fimbriae',
                                            Gene == 'new_gsp*' ~ 'General secreation pathway',
                                            Gene == 'new_ibeA' ~ 'Invasion of brain endothelium protein',
                                            Gene == 'new_iro*' ~ '',
                                            Gene == 'new_irp2/fyuA' ~ '',
                                            Gene == 'new_iutA/iuc*' ~ 'Aerobactin',
                                            Gene == 'new_kpsMT*' ~ 'K capsular protein',
                                            Gene == 'new_lhr' ~ 'Locus of Heat Resistance',
                                            Gene == 'new_mrk*' ~ '',
                                            Gene == 'new_mxi*' ~ 'membrane expression of invasion plasmid antigens',
                                            Gene == 'new_neu*' ~ '',
                                            Gene == 'new_nle*' ~ 'non-LEE-encoded (Nle) effector proteins',
                                            Gene == 'new_osp*' ~ 'Outer surface protein',
                                            Gene == 'new_pap*' ~ 'Pyelonephritis associated pilus',
                                            Gene == 'new_sep*' ~ 'locus of enterocyte effacement',
                                            Gene == 'new_sfa*' ~ 'S-fimbria-encoding determinent',
                                            Gene == 'new_shu*' ~ '',
                                            Gene == 'new_sit*' ~ 'Aerobactin',
                                            Gene == 'new_spa*' ~ 'Inner membrane protein',
                                            Gene == 'new_sub*' ~ 'Subtilase cytotoxin',
                                            Gene == 'new_tia' ~ '',
                                            Gene == 'new_tss*' ~ '',
                                            Gene == 'new_usp' ~ '',
                                            Gene == 'new_yag/ecp*' ~ 'E. coli common pilus',
                                            Gene == 'new_ybt*' ~ 'yersiniabactin'
                                                ))

vf_categories %>% mutate('Ref' = case_when(Gene == 'new_aaf*' ~ '10.1128/IAI.70.8.4302-4311.2002',
                                           Gene == 'new_aat*' ~ '10.1128/IAI.00513-09',
                                           Gene == 'new_afaE' ~ '10.1128/CMR.18.2.264-292.2005',
                                           Gene == 'new_bfp*' ~ '10.1111/j.1365-2958.1996.tb02491.x',
                                           Gene == 'new_ces*' ~ '10.1099/mic.0.26735-0',
                                           Gene == 'new_cfa*' ~ '10.1073/pnas.1808982115',
                                           Gene == 'new_cgs*' ~ '10.3390/ijms22189922',
                                           Gene == 'new_chu*' ~ '10.1046/j.1365-2958.1997.2641628.x',
                                           Gene == 'new_cjr/senB*' ~ '10.1128/JB.183.13.3958-3966.2001',
                                           Gene == 'new_clb*' ~ '10.1038/s41598-022-09274-x',
                                           Gene == 'new_cva*' ~ '10.1128/jb.179.20.6264-6270.1997',
                                           Gene == 'new_dra*' ~ '10.1128/IAI.00427-08',
                                           Gene == 'new_eit*' ~ '10.1128/genomeA.00863-16',
                                           Gene == 'new_eltA/B' ~ '10.1371/journal.pone.0123357',
                                           Gene == 'new_ent*' ~ '10.1128/jb.182.6.1768-1773.2000',
                                           Gene == 'new_esc*' ~ '10.1016/j.bbamem.2017.10.001',
                                           Gene == 'new_esp*' ~ '10.1073/pnas.191378598',
                                           Gene == 'new_fep*' ~ '10.1128/jb.169.8.3638-3646.1987',
                                           Gene == 'new_foc*' ~ '10.1128/jb.172.2.1114-1120.1990',
                                           Gene == 'new_gsp*' ~ '10.1128/jb.178.12.3544-3549.1996',
                                           Gene == 'new_ibeA' ~ '10.1128/IAI.03003-14',
                                           Gene == 'new_ipg*' ~ '10.3389/fmicb.2017.02390',
                                           Gene == 'new_iro*' ~ '10.1128/IAI.71.6.3285-3293.2003',
                                           Gene == 'new_irp2/fyuA' ~ '10.1007/s13205-021-02951-0',
                                           Gene == 'new_iutA/iuc*' ~ '10.1371/journal.pone.0057794',
                                           Gene == 'new_kpsMT*' ~ '10.1128/jb.173.15.4603-4610.1991',
                                           Gene == 'new_lhr' ~ '10.3389/fmicb.2020.00111',
                                           Gene == 'new_mrk*' ~ '10.1186/1471-2180-10-183',
                                           Gene == 'new_mxi*' ~ '10.1073/pnas.92.17.7996',
                                           Gene == 'new_neu*' ~ '10.1128/JB.186.3.706-712.2004',
                                           Gene == 'new_nle*' ~ '10.1099/mic.0.2006/003707-0',
                                           Gene == 'new_osp*' ~ '10.1128/mbio.01270-22',
                                           Gene == 'new_pap*' ~ '10.1128/iai.60.8.3416-3422.1992',
                                           Gene == 'new_sep*' ~ '10.1128/jb.182.22.6490-6498.2000',
                                           Gene == 'new_sfa*' ~ '10.1128/IAI.69.7.4248-4256.2001',
                                           Gene == 'new_shu*' ~ '10.1128/IAI.70.10.5503-5511.2002',
                                           Gene == 'new_sit*' ~ '10.1099/mic.0.28682-0',
                                           Gene == 'new_spa*' ~ '10.1016/j.femsle.2005.08.046',
                                           Gene == 'new_sub*' ~ '10.1007/s00204-020-02965-2',
                                           Gene == 'new_tia' ~ '10.1128/IAI.00613-17',
                                           Gene == 'new_tss*' ~ '10.15252/embj.2018100825',
                                           Gene == 'new_usp' ~ '10.1093/infdis/jit480',
                                           Gene == 'new_yag/ecp*' ~ '10.1074/jbc.M114.587717',
                                           Gene == 'new_ybt*' ~ '10.1038/nchembio.2441'
                                           
                                                ))

colnames(vf_df) <- gsub("new_", "", colnames(vf_df))

vf_categories <- vf_categories %>% arrange(Function, desc(Gene))

vf_categories <- vf_categories %>% mutate(Gene = str_replace(Gene, "new_", ""))

```

```{r}


# colnames(vf_df_source) <- gsub("(EC_custom.*)_[^_]+$","\\1", colnames(vf_df_source))
# colnames(vf_df_source) <- gsub("ABRICATE..EC_custom_","*", colnames(vf_df_source))
# colnames(vf_df_source) <- gsub("ABRICATE..vfdb_","", colnames(vf_df_source))


vf_df_total <- vf_df %>% select(-Consensus_phylogroup, -Revised_Source_Niche) %>% summarise(across(everything(), mean))

vf_df_total <- vf_df_total[, colSums(vf_df_total != 0) > 0.0001]

vf_df_total <- vf_df_total %>% select(any_of(vf_categories$Gene))

vf_df_total <- vf_categories %>% filter(Gene %in% colnames(vf_df_source)) %>% column_to_rownames('Gene') %>% select(Function)

vf_cols <- c('#dcbeff','#f032e6', '#e6194B', '#f58231', '#ffe119', '#3cb44b',  '#42d4f4', '#4363d8', 'darkblue','#000000', '#9A6324')

names(vf_cols) <- vf_annotation$Function %>% unique()

vf_cols <- list(Function = vf_cols)

vf_df_phylogroup <- vf_df %>% select(-Revised_Source_Niche) %>% group_by(Consensus_phylogroup) %>% summarise(across(everything(), mean)) %>% bind_rows(vf_df_total) %>% mutate(Consensus_phylogroup = if_else(is.na(Consensus_phylogroup), "Total", Consensus_phylogroup)) %>% filter(Consensus_phylogroup != "A") %>% column_to_rownames("Consensus_phylogroup")

vf_df_phylogroup <- vf_df_phylogroup %>% select(any_of(vf_categories$Gene))

vf_df_phylogroup <- vf_df_phylogroup[, colSums(vf_df_phylogroup != 0) > 0.1]

vf_phylogroup_w_total_heatmap <- vf_df_phylogroup %>% as.matrix() %>% pheatmap(color = colorRamp2(c(0, 1), c("white", "red")), legend = F, cluster_cols = F, fontsize_col = 8, cutree_rows = 3, display_numbers = T, cellwidth = 15, cellheight = 15, fontsize_number = 6, annotation_col = vf_annotation, annotation_colors = vf_cols, annotation_legend = F)


#colnames(vf_df_phylogroup) <- gsub("(EC_custom.*)_[^_]+$","\\1", colnames(vf_df_phylogroup))
#colnames(vf_df_phylogroup) <- gsub("ABRICATE..EC_custom_","*", colnames(vf_df_phylogroup))
#colnames(vf_df_phylogroup) <- gsub("ABRICATE..vfdb_","", colnames(vf_df_phylogroup))

vf_df_phylogroup <- vf_df %>% select(-Revised_Source_Niche) %>% group_by(Consensus_phylogroup) %>% summarise(across(everything(), mean)) %>% column_to_rownames("Consensus_phylogroup")

vf_df_phylogroup <- vf_df_phylogroup %>% select(any_of(vf_categories$Gene))

vf_df_phylogroup <- vf_df_phylogroup[, colSums(vf_df_phylogroup != 0) > 0.1]

vf_phylogroup_heatmap <- vf_df_phylogroup %>% as.matrix() %>% pheatmap(color = colorRamp2(c(0, 1), c("white", "red")), legend = F, cluster_cols = F, fontsize_col = 8, cutree_rows = 3, display_numbers = T, cellwidth = 15, cellheight = 15, fontsize_number = 6, annotation_col = vf_annotation, annotation_colors = vf_cols, annotation_legend = F)



vf_df_source <- vf_df %>% select(-Consensus_phylogroup) %>% group_by(Revised_Source_Niche) %>% summarise(across(everything(), mean)) %>% column_to_rownames("Revised_Source_Niche")

vf_df_source <- vf_df_source[, colSums(vf_df_source != 0) > 0.0001]

vf_df_source <- vf_df_source %>% select(any_of(vf_categories$Gene))

vf_annotation <- vf_categories %>% filter(Gene %in% colnames(vf_df_source)) %>% column_to_rownames('Gene') %>% select(Function)

vf_cols <- c('#dcbeff','#f032e6', '#e6194B', '#f58231', '#ffe119', '#3cb44b',  '#42d4f4', '#4363d8', 'darkblue','#000000', '#9A6324')

names(vf_cols) <- vf_annotation$Function %>% unique()

vf_cols <- list(Function = vf_cols)

vf_source_heatmap <- vf_df_source %>% as.matrix() %>% pheatmap(color = colorRamp2(c(0, 1), c("white", "red")), legend = F, annotation_legend = F, cluster_cols = F, fontsize_col = 8, cutree_rows = 3, display_numbers = T, cellwidth = 15, cellheight = 15, fontsize_number = 6, annotation_col = vf_annotation, annotation_colors = vf_cols)




vf_df_source %>% t() %>% pheatmap(color = colorRamp2(c(0, 1), c("white", "red")), legend = F, cluster_cols = F, fontsize_col = 8, cutree_rows = 3, display_numbers = T, cellwidth = 15, cellheight = 15, fontsize_number = 6, annotation_row = vf_annotation, annotation_colors = vf_cols)
```


```{r}


# colnames(vf_df_source) <- gsub("(EC_custom.*)_[^_]+$","\\1", colnames(vf_df_source))
# colnames(vf_df_source) <- gsub("ABRICATE..EC_custom_","*", colnames(vf_df_source))
# colnames(vf_df_source) <- gsub("ABRICATE..vfdb_","", colnames(vf_df_source))

vf_df_source <- vf_df %>% select(-Consensus_phylogroup) %>% group_by(Revised_Source_Niche) %>% summarise(across(everything(), mean)) %>% column_to_rownames("Revised_Source_Niche")

vf_df_source <- vf_df_source[, colSums(vf_df_source != 0) > 0.0001]

vf_source_heatmap <- vf_df_source %>% as.matrix() %>% pheatmap(color = colorRamp2(c(0, 1), c("white", "red")), legend = F, cluster_cols = T, fontsize_col = 5, cutree_rows = 3, cutree_cols = 2, display_numbers = T, cellwidth = 15, cellheight = 15, fontsize_number = 6)

d <- dist(t(vf_df_source), method = "euclidean")
hc1 <- hclust(d, method = "complete" )
plot(hc1)


#colnames(vf_df_phylogroup) <- gsub("(EC_custom.*)_[^_]+$","\\1", colnames(vf_df_phylogroup))
#colnames(vf_df_phylogroup) <- gsub("ABRICATE..EC_custom_","*", colnames(vf_df_phylogroup))
#colnames(vf_df_phylogroup) <- gsub("ABRICATE..vfdb_","", colnames(vf_df_phylogroup))

vf_df_phylogroup <- vf_df %>% select(-Revised_Source_Niche) %>% group_by(Consensus_phylogroup) %>% summarise(across(everything(), mean)) %>% column_to_rownames("Consensus_phylogroup")

vf_df_phylogroup <- vf_df_phylogroup %>% select(hc1$labels[hc1$order])

vf_df_phylogroup <- vf_df_phylogroup[, colSums(vf_df_phylogroup != 0) > 0.1]

vf_phylogroup_heatmap <- vf_df_phylogroup %>% as.matrix() %>% pheatmap(color = colorRamp2(c(0, 1), c("white", "red")), legend = F, cluster_cols = F, fontsize_col = 5, cutree_rows = 3, display_numbers = T, cellwidth = 15, cellheight = 15, fontsize_number = 6)




```

#Hybrid pathogen hunt
```{r}
genometa %>% rename_with(~str_replace(., "ABRICATE..vfdb_", ""), matches("ABRICATE")) %>% group_by(name) %>% mutate(has_stx = sum(stx1A, stx1B, stx2A, stx2B, stxA)) %>% filter(has_stx > 1) %>% select(name, ST_new,  Revised_Source_Niche, Collection_Year, Revised_Source_Type, Source_Details, matches("stx"), ColV, senB, ABRICATE..EC_custom_fyuA_Z35106.1, irp2, matches("cjr") ) %>% View()
```

```{r}
genometa <- vroom("delims/genometa_n5631.txt", show_col_types = FALSE)

#Read in our tree
tree <- ggtree::read.tree("analysis/TEA/cgmlst_coreugate/pad.nwk")

#Remove quotes from sample names
tree$tip.label <- gsub("'", "", tree$tip.label)

#Create a list of samples in our cohort
sample_names <- tree$tip.label %>% as.data.frame() %>% rename('name' = '.')


#Read in our cgMLST allelic distances
allelic_distances <- vroom(file = "analysis/TEA/cgmlst_coreugate/pad.txt", show_col_types = FALSE)

#Rename the name column
allelic_distances <- allelic_distances %>% rename(name = '5631')

#Change to long format
cgMLST_dists <- pivot_longer(allelic_distances, cols = 2:5632, names_repair = 'minimal')

#Rename name columns
colnames(cgMLST_dists) <- c('Sample 1', 'Sample 2', 'value')

if(!exists("genometa")){
        genometa <- vroom("delims/genometa_n5631.txt", show_col_types = FALSE)
}

#Create a dataframe we can bind to the distances
cgMLST_binder <- genometa %>% select(name, ColV, Revised_Source_Niche, Revised_Source_Type, Strain_MDR_status, Strain_CIA_status, clermontyping..clermontyping_phylogroup, Consensus_phylogroup, ST_new)

cgMLST_binder1 <- cgMLST_binder %>% rename(`Sample 1` = name)
cgMLST_binder2 <- cgMLST_binder %>% rename(`Sample 2` = name)

rm(cgMLST_binder)

cgMLST_dists2 <- cgMLST_dists %>% full_join(cgMLST_binder1, by = "Sample 1")

rm(cgMLST_dists)

cgMLST_dists2 <- cgMLST_dists2 %>% full_join(cgMLST_binder2, by = "Sample 2")

#Reorder columns (Sorry for the gibberish...)
cgMLST_dists2 <- cgMLST_dists2 %>% select(all_of(paste0(rev(colnames(cgMLST_binder1)[2:ncol(cgMLST_binder1)]),".x")), `Sample 1`, value, `Sample 2`, matches("\\.y"))

cgMLST_dists2 %>% filter(value <= 10) %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y) %>% select(`Sample 1`, `Sample 2`, value) -> overlaps
#Read in our mobtyper data
mobtyper_df <- vroom("analysis/TEA/mobtyper_results_concatenated.txt", show_col_types = FALSE)

#Filter in our mobtyper data and rename the sample ID column to name
mobtyper_df <- mobtyper_df %>% rename('name' = sample_id) %>% mutate(name = str_replace(name, ":.*", "")) %>% filter(name %in% tree$tip.label)

mobtyper_df %>% rename("Sample 1" = name) %>% right_join(overlaps, by = "Sample 1") -> overlaps_w_plas

mobtyper_df %>% rename("Sample 2" = name) %>% right_join(overlaps_w_plas, by = "Sample 2") -> overlaps_w_plas

genometa %>% select(name, ST_new, Revised_Source_Niche, Strain_CIA_status) %>% rename("Sample 1" = name) %>% right_join(overlaps_w_plas, by = "Sample 1") -> overlaps_w_plas

genometa %>% select(name, ST_new, Revised_Source_Niche, Strain_CIA_status) %>% rename("Sample 2" = name) %>% right_join(overlaps_w_plas, by = "Sample 2") -> overlaps_w_plas

overlaps_w_plas <- overlaps_w_plas %>% mutate("Source_Combo" = paste0(pmin(as.character(Revised_Source_Niche.x), as.character(Revised_Source_Niche.y)), "/", pmax(as.character(Revised_Source_Niche.x), as.character(Revised_Source_Niche.y))))

overlaps_w_plas <- overlaps_w_plas %>% mutate(sum_of_CIA = Strain_CIA_status.x + Strain_CIA_status.y)


df_new <- overlaps_w_plas %>% filter(primary_cluster_id.x == primary_cluster_id.y) %>% select(`rep_type(s).x`, secondary_cluster_id.x, primary_cluster_id.x, Revised_Source_Niche.y, `Sample 1`, value, `Sample 2`, Revised_Source_Niche.x, primary_cluster_id.y, secondary_cluster_id.y, `rep_type(s).y`, Source_Combo) %>% ungroup() %>% filter(grepl("IncF", `rep_type(s).x`), grepl("IncF", `rep_type(s).y`)) %>% unique() %>% filter(primary_cluster_id.x == primary_cluster_id.y) %>% group_by(primary_cluster_id.x, Source_Combo) %>% tally() %>% ungroup() %>% group_by(primary_cluster_id.x) %>% mutate(Total = sum(n)) %>% pivot_wider(names_from = Source_Combo, values_from = n, values_fill = 0) %>% ungroup() %>% select(!matches("Total"), Total) %>% column_to_rownames("primary_cluster_id.x")

df_new %>% pheatmap(display_numbers = T, number_format = "%.0f", color = colorRamp2(c(0, max(.)), c("white", "red")))


f_overlaps_w_plas <- overlaps_w_plas %>% filter(grepl("IncF", `rep_type(s).x`), grepl("IncF", `rep_type(s).y`)) %>% unique() %>% filter(primary_cluster_id.x == primary_cluster_id.y) 

plasmid_clusters <- read_delim("delims/plasmid_cluster_data.txt", delim = "\t")

cluster_cia_res_freq <- plasmid_clusters %>%
        group_by(primary_cluster_id) %>%
        mutate(Plasmid_CIA_status = replace_na(Plasmid_CIA_status, 0)) %>%
        summarise(generally_CIA = mean(Plasmid_CIA_status)) %>%
        mutate(generally_CIA = generally_CIA,
               generally_CIA_category = case_when(generally_CIA <= 0.33 ~ "Infrequent",
                                         generally_CIA >= 0.33 & generally_CIA <= 0.66 ~ "Moderate",
                                         generally_CIA >= 0.66 ~ "Predominant"))

f_overlaps_w_plas <- cluster_cia_res_freq %>% select(primary_cluster_id, generally_CIA, generally_CIA_category) %>% rename("primary_cluster_id.x" = primary_cluster_id) %>% right_join(f_overlaps_w_plas, by = "primary_cluster_id.x")
```

```{r}
library(highcharter)
library(htmlwidgets)
library(dplyr)

sankey1 <- f_overlaps_w_plas %>% select(generally_CIA_category, primary_cluster_id.x, Source_Combo, sum_of_CIA)

#sankey1 <- sankey1 %>% mutate(State = replace_na(State, "Unknown"))

sankey1 <- apply(sankey1, 2, FUN = function(x) paste0(x, " (",as.data.frame(table(x)[x])[,2],"/", as.character(nrow(sankey1)), ")"))

sankey1 <- as.data.frame(sankey1)

hchart(data_to_sankey(sankey1), "sankey", name = "Collection Breakdown")  %>%
        hc_plotOptions(series = list(dataLabels = list(style = list(fontSize = "20px"))))
```
```{r}

library(circlize)

cluster_ColV_freq <- plasmid_clusters %>%
        group_by(primary_cluster_id) %>%
        mutate(ColV = replace_na(ColV, 0)) %>%
        summarise(generally_ColV = mean(ColV)) %>%
        mutate(generally_ColV = generally_ColV,
               generally_ColV_category = case_when(generally_ColV <= 0.33 ~ "Infrequent",
                                         generally_ColV >= 0.33 & generally_ColV <= 0.66 ~ "Moderate",
                                         generally_ColV >= 0.66 ~ "Predominant"))

# Create an adjacency matrix: 
chord_data <- f_overlaps_w_plas %>% group_by(primary_cluster_id.x, Source_Combo) %>% summarise(counts = n()) %>%
        pivot_wider(id_cols = primary_cluster_id.x, names_from = Source_Combo, values_from = counts, values_fill = 0) %>% column_to_rownames("primary_cluster_id.x") %>% select(`Companion Animal/Livestock`, 1:2, `Companion Animal/Food`, 3:ncol(.)-1) %>% t() %>% as.data.frame() %>% select(AA297, AA171, AA337, AA336, AA315, AA172, AA176, AA747, AA175, AA179, AA319) %>% t() 

# Make the circular plot
chordDiagram(chord_data, transparency = 0.5)

# Clockwise labels may require increasing the canvas margins
circos.par(canvas.xlim=c(-1.5,1.5),canvas.ylim=c(-1.5,1.5))
#df.test=df[,c(1,2)]

set.seed(2)

circos.clear()

par(cex = 0.5)

grid.col = c(AA747 = "red", AA337 = "lightblue", AA336 = "darkgrey", AA319 = 'red4', AA315 = 'orange', AA297 = 'darkblue', AA179 = 'red3', AA176 = 'pink', AA175 = 'red2', AA172 = 'black', AA171 = 'blue')

chordDiagram(chord_data, annotationTrack = "grid", preAllocateTracks = 1, grid.col = grid.col)
circos.trackPlotRegion(track.index = 1, panel.fun = function(x, y) {
  xlim = get.cell.meta.data("xlim")
  ylim = get.cell.meta.data("ylim")
  sector.name = get.cell.meta.data("sector.index")
  circos.text(mean(xlim), ylim[1] + cm_h(1.5), sector.name, facing = "clockwise", niceFacing = TRUE, adj = c(0, 0.5), cex =1.5)
  circos.axis(h = "top", labels.cex =1, major.tick.percentage = 0.2, sector.index = sector.name, track.index = 2)
}, bg.border = NA)
```


### Plasmids in absence of cgMLST overlap

```{r}
genometa <- vroom("delims/genometa_n5631.txt", show_col_types = FALSE)

#Read in our tree
tree <- ggtree::read.tree("analysis/TEA/cgmlst_coreugate/pad.nwk")

#Remove quotes from sample names
tree$tip.label <- gsub("'", "", tree$tip.label)

#Create a list of samples in our cohort
sample_names <- tree$tip.label %>% as.data.frame() %>% rename('name' = '.')


#Read in our cgMLST allelic distances
allelic_distances <- vroom(file = "analysis/TEA/cgmlst_coreugate/pad.txt", show_col_types = FALSE)

#Rename the name column
allelic_distances <- allelic_distances %>% rename(name = '5631')

#Change to long format
cgMLST_dists <- pivot_longer(allelic_distances, cols = 2:5632, names_repair = 'minimal')

#Rename name columns
colnames(cgMLST_dists) <- c('Sample 1', 'Sample 2', 'value')

if(!exists("genometa")){
        genometa <- vroom("delims/genometa_n5631.txt", show_col_types = FALSE)
}

#Create a dataframe we can bind to the distances
cgMLST_binder <- genometa %>% select(name, ColV, Revised_Source_Niche, Revised_Source_Type, Strain_MDR_status, Strain_CIA_status, clermontyping..clermontyping_phylogroup, Consensus_phylogroup, ST_new)

cgMLST_binder1 <- cgMLST_binder %>% rename(`Sample 1` = name)
cgMLST_binder2 <- cgMLST_binder %>% rename(`Sample 2` = name)

rm(cgMLST_binder)

cgMLST_dists2 <- cgMLST_dists %>% full_join(cgMLST_binder1, by = "Sample 1")

rm(cgMLST_dists)

cgMLST_dists2 <- cgMLST_dists2 %>% full_join(cgMLST_binder2, by = "Sample 2")

#Reorder columns (Sorry for the gibberish...)
cgMLST_dists2 <- cgMLST_dists2 %>% select(all_of(paste0(rev(colnames(cgMLST_binder1)[2:ncol(cgMLST_binder1)]),".x")), `Sample 1`, value, `Sample 2`, matches("\\.y"))

cgMLST_dists2  %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y) %>% select(`Sample 1`, `Sample 2`, value) -> overlaps
#Read in our mobtyper data
mobtyper_df <- vroom("analysis/TEA/mobtyper_results_concatenated.txt", show_col_types = FALSE)

#Filter in our mobtyper data and rename the sample ID column to name
mobtyper_df <- mobtyper_df %>% rename('name' = sample_id) %>% mutate(name = str_replace(name, ":.*", "")) %>% filter(name %in% tree$tip.label)

#Define major virulence clusters
major_virulence_clusters <- c('AA176', 'AA315', 'AA179', 'AA175', 'AA174', 'AA638', 'AA352', 'AA337', 'AA171', 'AA324', 'AA297', 'AA735')

#Filter to our list of plasmids which have a median virulence gene count of over 1 and are present at counts of > 100
major_f_plasmids <- mobtyper_df %>% filter(primary_cluster_id %in% major_virulence_clusters)

vir_plas_carrying_overlaps <- overlaps %>% filter(`Sample 1` %in% major_f_plasmids$name )

vir_plas_carrying_overlaps <- major_f_plasmids %>% rename("Sample 1" = name) %>% right_join(vir_plas_carrying_overlaps, by = "Sample 1")

vir_plas_carrying_overlaps <- major_f_plasmids %>% rename("Sample 2" = name) %>% right_join(vir_plas_carrying_overlaps, by = "Sample 2")

major_f_plasmids_cross_source <- vir_plas_carrying_overlaps %>% filter(primary_cluster_id.x == primary_cluster_id.y)

genometa %>% select(name, ST_new, Revised_Source_Niche, Strain_CIA_status) %>% rename("Sample 1" = name) %>% right_join(major_f_plasmids_cross_source, by = "Sample 1") -> major_f_plasmids_cross_source

genometa %>% select(name, ST_new, Revised_Source_Niche, Strain_CIA_status) %>% rename("Sample 2" = name) %>% right_join(major_f_plasmids_cross_source, by = "Sample 2") -> major_f_plasmids_cross_source

major_f_plasmids_cross_source <- major_f_plasmids_cross_source %>% mutate("Source_Combo" = paste0(pmin(as.character(Revised_Source_Niche.x), as.character(Revised_Source_Niche.y)), "/", pmax(as.character(Revised_Source_Niche.x), as.character(Revised_Source_Niche.y))))

major_f_plasmids_cross_source <- major_f_plasmids_cross_source %>% mutate(sum_of_CIA = Strain_CIA_status.x + Strain_CIA_status.y)


df_new <- major_f_plasmids_cross_source %>% filter(primary_cluster_id.x == primary_cluster_id.y) %>% select(`rep_type(s).x`, secondary_cluster_id.x, primary_cluster_id.x, Revised_Source_Niche.y, `Sample 1`, value, `Sample 2`, Revised_Source_Niche.x, primary_cluster_id.y, secondary_cluster_id.y, `rep_type(s).y`, Source_Combo) %>% ungroup() %>% unique() %>% group_by(primary_cluster_id.x, Source_Combo) %>% tally() %>% ungroup() %>% group_by(primary_cluster_id.x) %>% mutate(Total = sum(n)) %>% pivot_wider(names_from = Source_Combo, values_from = n, values_fill = 0) %>% ungroup() %>% select(!matches("Total"), Total) %>% column_to_rownames("primary_cluster_id.x")

df_new %>% pheatmap(display_numbers = T, number_format = "%.0f", color = colorRamp2(c(0, max(.)), c("white", "red")))

circos.clear()

df_new_less_total <- df_new %>% select(-Total)

circos.par(canvas.xlim=c(-1.5,1.5),canvas.ylim=c(-1.5,1.5))

set.seed(1)

chordDiagram(df_new_less_total, annotationTrack = "name", preAllocateTracks = 1)
circos.trackPlotRegion(track.index = 1, panel.fun = function(x, y) {
  xlim = get.cell.meta.data("xlim")
  ylim = get.cell.meta.data("ylim")
  sector.name = get.cell.meta.data("sector.index")
  circos.text(mean(xlim), ylim[1] + cm_h(1.5), sector.name, facing = "clockwise", niceFacing = TRUE, adj = c(0, 0.5), cex =1.5)
  circos.axis(h = "top", labels.cex =1, major.tick.percentage = 0.2, sector.index = sector.name, track.index = 2)
}, bg.border = NA)
```
